---
title: '`r params[["title"]]`'
author:
- name: '`r params[["author_name"]]`'
  affiliation: '`r params[["author_affiliation"]]`'
  email: '`r params[["author_email"]]`'
date: '`r format(Sys.time(), "%B %d, %Y")`'
monofont: "Source Code Pro"
monofontoptions: "Scale=0.7"
params:
  input_directory: NULL
  output_directory: NULL
  title: "Post-Imputation Quality-Control"
  author_name: "Firstname Lastname"
  author_affiliation: "Institution"
  author_email: "some@email.com"
  cache: FALSE
  show_code: FALSE
  n_cores: 1
  dpi: 120
  gg_fontsize: 12
output:
  bookdown::html_document2:
    theme: simplex
    toc: true
    toc_depth: 3
    toc_float: 
      collapsed: false
    fig_width: 6.3
    fig_height: 4.7
    number_sections: true
    self_contained: true
    mathjax: default
    df_print: kable
---

```{r setup, include = FALSE}
options(stringsAsFactors = FALSE)

### Load packages and functions
library(parallel)
library(tidyverse)
library(scales)
library(grid)

pretty_kable <- function (
  data, 
  font_size = 12, 
  format_args = list(scientific = -1, digits = 3, big.mark = ","), 
  col.names = NA,
  full_width = TRUE,
  format = "html",
  ...
) {
  output <- knitr::kable(
    x = data, 
    format.args = format_args, 
    col.names = col.names,
    ...
  )
  kableExtra::kable_styling(
    kable_input = output,
    bootstrap_options = c("striped", "hover", "condensed", "responsive"),
    full_width = full_width,
    position = "center",
    font_size = font_size
  )
}


### Set knitr rmarkdown chunk options
dir.create(
  path = paste0(params[["output_directory"]], "/cache/"), 
  recursive = TRUE, 
  showWarnings = FALSE, 
  mode = "0777"
)
knitr::opts_chunk$set(
  results = "asis",
  size = "small",
  include = TRUE,
  echo = params[["show_code"]],
  warning = params[["show_code"]],
  message = params[["show_code"]],
  dpi = params[["dpi"]],
  tidy = FALSE,
  crop = TRUE,
  autodep = TRUE,
  fig.align = "center",
  fig.pos = "!H",
  cache = params[["cache"]],
  cache.path = paste0(params[["output_directory"]], "/cache/"),
  fig.path = paste0(params[["output_directory"]], "/cache/")
)


### Define theme
theme_set(theme_light(base_size = params[["gg_fontsize"]]))
```

# Method

Imputation performed using the [Sanger Imputation Services Pipeline](https://imputation.sanger.ac.uk/).

## Haplotype Reference Consortium (release 1.1)

### Description

The aim of the [Haplotype Reference Consortium (HRC)](http://www.haplotype-reference-consortium.org/) is to create a large reference panel of human haplotypes by combining together sequencing data from multiple cohorts.  
This HRC.r1 release consists of 64,940 haplotypes of predominantly European ancestry.  
More information about the HRC can be found [here](http://www.haplotype-reference-consortium.org/).

### Reference panel information

* Chromosomes: 1-22 and X
* VCF lines: 40,405,505 (1,273,927 chrX)
* Biallelic SNPs: 40,176,563 (1,263,515 chrX)
* Tri-allelic SNPs: 114,183 (5,197 chrX)
* Quad-allelic SNPs: 192 (6 chrX)
* Samples: (31,500 on chrX)
* Haplotyes: (63,000/49,248 on chrX PAR/nonPAR)
* Ethnicity: Mostly pan European + 1000 Genomes Phase 3

## Pre-phasing and imputation with `SHAPEIT2+PBWT`

Prior to imputation with PBWT, each chromosome is phased against the chosen reference panel using SHAPEIT2 (v2.r790).  
The phasing does not take into account any sources of information other than the input genotypes when phasing. 

## Output files

The format of the returned data will be in the [Variant Call Format (VCF)](http://samtools.github.io/hts-specs/VCFv4.2.pdf).  
The returned files will be:

1. Imputation pipelines: One VCF per-chromosome for the PBWT imputation pipelines.
1. Each VCF will be have been compressed with `bgzip` and have an associated `tabix` index (`.tbi`) file allowing random access queries with bcftools, tabix and other VCF processing tools.
1. A tarball of the various log files associated with the pipeline is also returned.

The VCF header should contain up-to-date information about the meaning of the INFO and FORMAT tags as well as program versions and command lines used in the pipeline.  
A quick way to view the header is with `bcftools view -h in.vcf.gz` or `tabix -H in.vcf.gz`.

Site specific metrics are encoded in the INFO column of the VCF.  
Depending on the pipeline choice, these may include:

* __INFO/AC__: non-reference allele count across called genotypes.
* __INFO/AN__: total number of alleles in called genotypes.
* __INFO/RefPanelAF__: non reference allele frequency in the imputation reference panel.
* __INFO/INFO__: IMPUTE2 info score as calculated by the `bcftools +impute-info` plugin from the posterior genotype probabilities (FORMAT/GP).

Genotype or haplotype specific metrics are encoded in the __FORMAT__ columns of the VCF.  
Depending on the pipeline choice, these may include:

* __FORMAT/GT__: the genotype or haplotype call. Original genotype calls will not be altered.
* __FORMAT/DS__: genotype dosage.
* __FORMAT/GP__: the posterior genotype probabilities.

See the [resources](https://imputation.sanger.ac.uk/?resources=1) page for help converting phased output to Oxford HAP-SAMPLE format or imputed output to Oxford GEN-SAMPLE format.

For imputation pipelines, the __ID__ column will be filled with [dbSNP rsIDs](http://www.ncbi.nlm.nih.gov/SNP/).  
See the VCF header lines for the dbSNP version used.

## Citation

* The [Sanger Imputation Service](https://imputation.sanger.ac.uk/) was introduced as part of the first release of the [Haplotype Reference Consortium](http://www.haplotype-reference-consortium.org/). 
    * S. McCarthy et al. (2016) A reference panel of 64,976 haplotypes for genotype imputation, Nature Genetics. 48(10):1279-83. doi:10.1038/ng.3643
* The HRC.r1 reference panel is described in the following paper.  
    * S. McCarthy et al. (2016) A reference panel of 64,976 haplotypes for genotype imputation, Nature Genetics. 48(10):1279-83. doi:10.1038/ng.3643
* [SHAPEIT2](http://shapeit.fr/) is used for pre-phasing:
    * O. Delaneau, J. Marchini, JF. Zagury (2012) A linear complexity phasing method for thousands of genomes. Nature Methods. 9(2):179-81. doi:10.1038/nmeth.1785
* [PBWT](https://github.com/richarddurbin/pbwt) is used for imputation, the imputation algorithm used by PBWT is not yet published. However you may cite the original PBWT paper about the storage format and haplotype matching algorithm:
    * R. Durbin (2014) Efficient haplotype matching and storage using the positional Burrows-Wheeler transform (PBWT), Bioinformatics. 30(9):1266-72. doi:10.1093/bioinformatics/btu014

# Imputation report

```{r read, include = FALSE}
list_vcf <- list.files(path = params[["data_directory"]], pattern = "vcf.gz$", full.names = TRUE)
all_chrosomoses <- lapply(
  X = list_vcf,
  FUN = function(ichr) {
    data.table::fread(
      file = ichr, 
      header = TRUE, 
      check.names = FALSE,
      skip = "#C",
      select = c("#CHROM", "POS", "INFO"), 
      colClasses = c("#CHROM" = "character", "POS" = "integer", "INFO" = "character"),
      showProgress = FALSE
    ) %>% 
      dplyr::rename(CHR = `#CHROM`) %>% 
      tidyr::separate(col = "INFO", into = c("RefPanelAF", "AC", "AN", "INFO"), sep = ";") %>% 
      dplyr::mutate_at(.vars = dplyr::vars(c("RefPanelAF", "AC", "AN", "INFO")), .funs = ~as.numeric(gsub(".*=", "", .x))) %>% 
      dplyr::mutate(
        # CHR = as.character(CHR),
        AF = AC / AN,
        bin_af = cut(AF, breaks = seq(0, 1, 0.1), include.lowest = TRUE),
        bin_info = cut(INFO, breaks = seq(0, 1, 0.1), include.lowest = TRUE),
        bin_af = forcats::fct_explicit_na(bin_af),
        bin_info = forcats::fct_explicit_na(bin_info)
      )
  }
)
names(all_chrosomoses) <- gsub("^([^.]+)\\..*", "\\1", basename(list_vcf))
```

## Genome-Wide

```{r genomewide}
cat("### Variants binned by alternate allele frequency\n\n")
all_chrosomoses %>% 
  dplyr::bind_rows() %>% 
  dplyr::count(bin_af) %>% 
  dplyr::mutate(p = n / sum(n)) %>% 
  dplyr::add_row(bin_af = "Total", n = sum(.$n), p = 1) %>%
  dplyr::mutate(
    n = scales::comma(n),
    p = scales::percent(p, accuracy = 0.0001)
  ) %>% 
  dplyr::rename(
    `Alternate Allele Frequency` = bin_af,
    Count = n,
    `%` = p
  ) %>% 
  pretty_kable(
    align = "c", 
    table.attr = "style='width:65%;'"
  )
cat("\n\n")
cat("### Variants binned by imputation INFO score\n\n")
all_chrosomoses %>% 
  dplyr::bind_rows() %>% 
  dplyr::count(bin_info) %>% 
  dplyr::mutate(p = n / sum(n)) %>% 
  dplyr::add_row(bin_info = "Total", n = sum(.$n), p = 1) %>%
  dplyr::mutate(
    n = scales::comma(n),
    p = scales::percent(p, accuracy = 0.0001)
  ) %>% 
  dplyr::rename(
    `INFO score` = bin_info,
    Count = n,
    `%` = p
  ) %>% 
  pretty_kable(
    align = "c", 
    table.attr = "style='width:65%;'"
  )
cat("\n\n")
```

## Chromosome

```{r chromosome}
for (ichr in c(1:22, "X")) {
  cat("### Chromosome ", ichr, "\n\n", sep = "")
  tmp <- all_chrosomoses[[ichr]]
  
  cat("#### Alternate allele frequency in cohort vs HRC\n\n")
  p1 <- ggplot2::ggplot(
    data = tmp, 
    mapping = ggplot2::aes(x = RefPanelAF, y = AF, colour = INFO > 0.8)
  ) +
    ggplot2::geom_point(size = 0.1, na.rm = TRUE) +
    ggplot2::scale_colour_viridis_d(begin = 0.2, end = 0.8, na.value = "firebrick2") +
    ggplot2::labs(x = "HRC alternate allele frequency", y = "Alternate allele frequency") +
    ggplot2::scale_x_continuous(labels = scales::percent_format(accuracy = 0.01), expand = c(0, 0)) +
    ggplot2::scale_y_continuous(labels = scales::percent_format(accuracy = 0.01), expand = c(0, 0))
  print(p1)
  cat("\n\n")
  
  cat("#### Variants binned by alternate allele frequency\n\n")
  tmp %>% 
    dplyr::count(bin_af) %>% 
    dplyr::mutate(p = n / sum(n)) %>% 
    dplyr::add_row(bin_af = "Total", n = sum(.$n), p = 1) %>%
    dplyr::mutate(
      n = scales::comma(n),
      p = scales::percent(p, accuracy = 0.0001)
    ) %>% 
    dplyr::rename(
      `Alternate Allele Frequency` = bin_af,
      Count = n,
      `%` = p
    ) %>% 
    pretty_kable(
      align = "c", 
      table.attr = "style='width:65%;'"
    )
  cat("\n\n")
  p2 <- ggplot2::ggplot(
    data = tmp %>% 
      dplyr::count(bin_af) %>% 
      dplyr::mutate(p = scales::percent(n / sum(n), accuracy = 0.01)), 
    mapping = ggplot2::aes(x = bin_af, y = n, label = p)
  ) +
    ggplot2::geom_bar(fill = scales::viridis_pal(begin = 0.5, end = 0.5)(1), stat = "identity") +
    ggplot2::geom_text(vjust = -0.1) +
    ggplot2::scale_y_continuous(
      labels = scales::comma, 
      expand = ggplot2::expand_scale(mult = c(0, 0.10))
    ) +
    ggplot2::labs(x = "Alternate allele frequency", y = "Count") +
    ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 45, hjust = 1))
  print(p2)
  cat("\n\n")
  
  cat("#### Variants binned by imputation INFO score\n\n")
  tmp %>% 
    dplyr::count(bin_info) %>% 
    dplyr::mutate(p = n / sum(n)) %>% 
    dplyr::add_row(bin_info = "Total", n = sum(.$n), p = 1) %>%
    dplyr::mutate(
      n = scales::comma(n),
      p = scales::percent(p, accuracy = 0.0001)
    ) %>% 
    dplyr::rename(
      `INFO score` = bin_info,
      Count = n,
      `%` = p
    ) %>% 
    pretty_kable(
      align = "c", 
      table.attr = "style='width:65%;'"
    )
  cat("\n\n")
  p3 <- ggplot2::ggplot(
    data = tmp %>% 
      dplyr::count(bin_info) %>% 
      dplyr::mutate(p = scales::percent(n / sum(n), accuracy = 0.01)), 
    mapping = ggplot2::aes(x = bin_info, y = n, label = p)
  ) +
    ggplot2::geom_bar(fill = scales::viridis_pal(begin = 0.5, end = 0.5)(1), stat = "identity") +
    ggplot2::geom_text(vjust = -0.1) +
    ggplot2::scale_y_continuous(
      labels = scales::comma, 
      expand = ggplot2::expand_scale(mult = c(0, 0.10))
    ) +
    ggplot2::labs(x = "INFO score", y = "Count") +
    ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 45, hjust = 1))
  print(p3)
  cat("\n\n")
  
  cat("#### Manhattan plot of the INFO score across the chromosome\n\n")
  p4 <- ggplot2::ggplot(
    data = tidyr::drop_na(tmp, INFO), 
    mapping = ggplot2::aes(x = POS / 1e6, y = INFO, colour = INFO > 0.8)
  ) +
    ggplot2::geom_point(size = 0.1, na.rm = TRUE) +
    ggplot2::scale_colour_viridis_d(begin = 0.2, end = 0.8) +
    ggplot2::scale_x_continuous(labels = scales::comma) +
    ggplot2::scale_y_continuous(expand = c(0, 0)) +
    ggplot2::labs(x = "Position (Mb)", y = "INFO score")
  print(p4)
  cat("\n\n")
  
  cat("#### Check for consistent ascending chromosome/position ordering\n\n")
  p5 <- ggplot2::ggplot(
    data = tidyr::drop_na(tmp, INFO),
    mapping = ggplot2::aes(x = POS / 1e6, y = seq_along(POS) / 1e3)
  ) +
    ggplot2::geom_point(
      size = 0.1, 
      colour = scales::viridis_pal(begin = 0.5, end = 0.5)(1), 
      na.rm = TRUE
    ) +
    ggplot2::scale_x_continuous(labels = scales::comma) +
    ggplot2::scale_y_continuous(labels = scales::comma) +
    ggplot2::labs(x = "Position on chromosome (Mb)", y = "Line number (x 1,000)")
  print(p5)
  cat("\n\n")
}
```


# R session information

```{r session-info, results = "markup"}
options("width" = 110)
sessioninfo::session_info()
```
