---
title: '`r params[["title"]]`'
author:
- name: '`r params[["author_name"]]`'
  affiliation: '`r params[["author_affiliation"]]`'
  email: '`r params[["author_email"]]`'
date: '`r format(Sys.time(), "%B %d, %Y")`'
monofont: "Source Code Pro"
monofontoptions: "Scale=0.7"
params:
  input_directory: NULL
  output_directory: NULL
  cohort_name: NULL
  array: NULL
  callrate_samples: 0.95
  callrate_snps: 0.95
  heterozygosity_treshold: 4
  maf_threshold: 0.01
  hwe_pvalue: 0.0001
  includes_relatives: FALSE
  mendelian_samples: 0.05
  mendelian_snp: 0.1
  IBD_threshold: 0.2
  population: NULL
  pca_components: 10
  pca_threshold: 3
  check_bim_script: "HRC-1000G-check-bim.pl"
  ref1kg_panel: NULL
  ref1kg_population: NULL
  ref1kg_genotypes: NULL
  ref1kg_legend: NULL
  ref1kg_fasta: NULL
  bin_path: !r list(bcftools = "/usr/bin/bcftools", bgzip = "/usr/bin/bgzip", plink = "/usr/bin/plink1.9", gcta = "/usr/bin/gcta64")
  title: "Genomic Array Quality-Control"
  author_name: "Firstname Lastname"
  author_affiliation: "Institution"
  author_email: "some@email.com"
  cache: FALSE
  show_code: FALSE
  n_cores: 1
  dpi: 120
  gg_fontsize: 12
output:
  bookdown::html_document2:
    theme: simplex
    toc: true
    toc_depth: 3
    toc_float: 
      collapsed: false
    fig_width: 6.3
    fig_height: 4.7
    fig_caption: true
    number_sections: true
    self_contained: true
    df_print: kable
editor_options: 
  chunk_output_type: inline
---

```{r setup, include = FALSE, echo = FALSE}
options(stringsAsFactors = FALSE, knitr.table.format = "html")

invisible(
  sapply(
    X = paste(params[["output_directory"]], c("tmp", "plink_qc", "vcf_qc"), sep = "/"), 
    FUN = dir.create,
    showWarnings = FALSE, 
    recursive = TRUE, 
    mode = "0777"
  )
)
temp_directory <- paste0(params[["output_directory"]], "/tmp")
dir.create(paste0(temp_directory, "/1kgenome_small"), showWarnings = FALSE, recursive = TRUE, mode = "0777")

if (params[["cache"]]) {
  dir.create(paste0(params[["output_directory"]], "/cache/"), recursive = TRUE, showWarnings = FALSE, mode = "0777")
}

knitr::opts_chunk$set(
  results = "asis",
  size = "small",
  include = TRUE,
  echo = params[["show_code"]],
  warning = params[["show_code"]],
  message = params[["show_code"]],
  dpi = params[["dpi"]],
  tidy = FALSE,
  crop = TRUE,
  autodep = TRUE,
  fig.align = "center",
  fig.pos = "!H",
  cache = params[["cache"]],
  cache.path = paste0(params[["output_directory"]], "/cache/"),
  fig.path = paste0(params[["output_directory"]], "/figures/")
)

pretty_kable <- function(
  data, 
  font_size = 12, 
  format_args = list(scientific = -1, digits = 3, big.mark = ","), 
  col.names = NA,
  full_width = TRUE,
  echo = TRUE,
  ...
) {
  colnames(data) <- Hmisc::capitalize(colnames(data))
  options(knitr.table.format = "html")
  output <- knitr::kable(x = data, format.args = format_args, col.names = col.names, ...)
  output <- kableExtra::kable_styling(
    kable_input = output,
    bootstrap_options = c("striped", "hover", "condensed", "responsive"),
    full_width = ifelse(is.null(full_width), FALSE, full_width),
    position = "center",
    font_size = font_size
  )
  
  if (echo) print(output) else invisible(output)
}

pval_trans <- function() {
  neglog10_breaks <- function(n = 5) {
    function(x) {
      rng <- -log(range(x, na.rm = TRUE), base = 10)
      min <- 0
      max <- floor(rng[1])
      if (max == min) {
        10^-min
      } else {
        by <- floor((max - min) / n) + 1
        10^-seq(min, max, by = by)
      }
    }
  }
  scales::trans_new(
    name = "pval",
    transform = function(x) {-log(x, 10)},
    inverse = function(x) {10^-x},
    breaks = neglog10_breaks(),
    format = function(x) {
      parse(
        text = gsub("e", " %*% 10^", gsub("1e+00", "1", scales::scientific_format()(x), fixed = TRUE))
      )
    },
    domain = c(0, 1)
  )
}

`%>%` <- magrittr::`%>%`

ggplot2::theme_set(ggplot2::theme_light(base_size = params[["gg_fontsize"]]))
```

```{r data}
cat(
  "Data from:", dirname(params[["input_directory"]]), 
  file = paste0(params[["output_directory"]], "/README.md")
)

system(command = paste(
  params[["bin_path"]][["plink"]], 
  "--bfile", params[["input_directory"]], 
  "--snps-only", 
  "--make-bed",
  "--out", paste0(temp_directory, "/raw_data")
), ignore.stdout = TRUE, ignore.stderr = TRUE)
```



# Parameters

* Raw data: 
    `` `r params[["input_directory"]]` ``
* 1,000 Genomes files: 
    `` `r params[["ref1kg_genotypes"]]` ``
* Reference population: 
    `` `r if (is.null(params[["population"]])) "unspecified" else params[["population"]]` ``
* Sample-based call rate threshold: 
    `` `r params[["callrate_samples"]]` ``
* Variant-based call rate threshold: 
    `` `r params[["callrate_snps"]]` ``
* Number of standard deviations (SD) from the mean heterozygosity rate: 
    `` `r params[["heterozygosity_treshold"]]` ``
* Proportion of identity by descent (IBD) threshold: 
    `` `r params[["IBD_threshold"]]` ``
* Number of components from the principal component analysis (PCA): 
    `` `r params[["pca_components"]]` ``
* Number of standard deviations (SD) from the population centroid based on the PCA: 
    `` `r params[["pca_threshold"]]` ``
* Minor allele frequency (MAF) threshold: 
    `` `r params[["maf_threshold"]]` ``
* Hardy-Weinberg equilibrium pvalue: 
    `` `r scales::scientific(params[["hwe_pvalue"]])` ``
* Mendelian error rate threshold:
    + Max per-trio error rate: 
        `` `r if (params[["includes_relatives"]]) params[["mendelian_samples"]] else "not applicable"` `` 
    + Max per-variant error rate: 
        `` `r if (params[["includes_relatives"]]) params[["mendelian_snp"]] else "not applicable"` `` 



# Cohort overview

Genotyped using `r if (is.null(params[["array"]])) "unspecified" else params[["array"]]` array.

```{r cohort_overview, include = FALSE}
samples <- dplyr::mutate_at(
  .tbl = utils::read.table(paste0(temp_directory, "/raw_data.fam")), 
  .vars = 1:2, 
  .funs = as.character
)

samples_duplicated <- table(factor(
  x = table(gsub("[_-][^-_]*$", "", samples[[2]]))==2,
  levels = c(FALSE, TRUE)
))[["TRUE"]]

snps_number <- nrow(data.table::fread(input = paste0(temp_directory, "/raw_data.bim"), select = 1))

cohort_details <- paste0(dirname(params[["input_directory"]]), "/Details_Cohorts.xlsx")
if (file.exists(cohort_details)) {
  recruit_centre <- readxl::read_xlsx(path = cohort_details, sheet = 1) %>% 
    dplyr::select(Sample.Name = 1, Cohort = 2, Gender = 3) %>%
    dplyr::filter(Sample.Name %in% samples[[2]]) %>% 
    dplyr::mutate(Sample.Name = as.character(Sample.Name))
} else {
  recruit_centre <- samples %>%
    dplyr::select(Sample.Name = 2, Gender = 5) %>%
    dplyr::mutate(Cohort = !!params[["cohort_name"]]) %>%
    dplyr::mutate(Sample.Name = as.character(Sample.Name))
}

recruit_centre_table <- recruit_centre %>%
  dplyr::mutate(Gender = factor(Gender, levels = c(1, 2, 0))) %>% 
  dplyr::group_by(Cohort, Gender, .drop = FALSE) %>% 
  dplyr::tally() %>%
  tidyr::spread(Gender, n) %>% 
  dplyr::rowwise() %>% 
  dplyr::mutate(Total = sum(`0`, `1`, `2`)) %>% 
  dplyr::ungroup() %>% 
  dplyr::rename(
    male = `1`,
    female = `2`,
    unspecified = `0`
  ) %>%
  dplyr::add_row(
    Cohort = "ALL", 
    Total = sum(.$Total),
    male = sum(.$male),
    female = sum(.$female),
    unspecified = sum(.$unspecified)
  )
```

PLINK files includes:

* **`r scales::comma(nrow(recruit_centre))`** samples  
    + Note: **`r scales::comma(samples_duplicated)`** samples seemed to have duplicates
* **`r scales::comma(snps_number)`** variants

```{r cohort-overview-table}
pretty_kable(
  data = recruit_centre_table,
  align = "c", 
  table.attr = "style='width:65%;'",
  caption = "Number of samples per cohorts.", 
  echo = FALSE
) %>%
  kableExtra::row_spec(row = nrow(recruit_centre_table), bold = TRUE) %>%
  kableExtra::add_header_above(c(" " = 1, "Gender" = 3, " " = 1))
```



# Sample-based QC {#sample-qc}


## Sample call rate

```{r sample-call-rate-01, include = FALSE}
system(command = paste(
  params[["bin_path"]][["plink"]], 
  "--bfile", paste0(temp_directory, "/raw_data"), 
  "--missing", 
  "--out", paste0(temp_directory, "/raw_data")
), ignore.stdout = TRUE, ignore.stderr = TRUE)

callrate_samples <- data.table::fread(paste0(temp_directory, "/raw_data.imiss")) %>%
  dplyr::mutate_at(.vars = c("FID", "IID"), as.character) %>% 
  dplyr::arrange(F_MISS) %>%
  dplyr::mutate(labs = ifelse(F_MISS > 1 - params[["callrate_samples"]], IID, NA))

samples_toexclude <- callrate_samples %>%
  dplyr::filter(!is.na(labs)) %>%
  dplyr::select(FID, IID) %>%
  dplyr::mutate(QC = "Sample_Call_Rate")

system(command = paste(
  params[["bin_path"]][["plink"]], 
  "--bfile ", paste0(temp_directory, "/raw_data"), 
  "--freq", 
  "--out", paste0(temp_directory, "/raw_data")
), ignore.stdout = TRUE, ignore.stderr = TRUE)

freqsnp <- data.table::fread(paste0(temp_directory, "/raw_data.frq"))

freqsnp %>% 
  dplyr::filter(MAF >= params[["maf_threshold"]]) %>% 
  .[["SNP"]] %>% 
  utils::write.table(
    col.names = FALSE, 
    row.names = FALSE, 
    quote = FALSE,
    file = paste0(temp_directory, "/raw_data_maf_geq.txt")
  )

snp_maf_lower <- freqsnp %>% 
  dplyr::filter(MAF < params[["maf_threshold"]]) %>% 
  .[["SNP"]]

is_maf_lower <- length(snp_maf_lower) > 0

if (is_maf_lower) {
  utils::write.table(
    snp_maf_lower, 
    col.names = FALSE, row.names = FALSE, quote = FALSE,
    file = paste0(temp_directory, "/raw_data_maf_lower.txt")
  )
}

system(command = paste(
  params[["bin_path"]][["plink"]], 
  "--bfile", paste0(temp_directory, "/raw_data"), 
  "--missing", 
  "--extract", paste0(temp_directory, "/raw_data_maf_geq.txt"), 
  "--out", paste0(temp_directory, "/raw_data_maf_geq")
), ignore.stdout = TRUE, ignore.stderr = TRUE)

if (is_maf_lower) {
  system(command = paste(
    params[["bin_path"]][["plink"]], 
    "--bfile", paste0(temp_directory, "/raw_data"), 
    "--missing", 
    "--extract", paste0(temp_directory, "/raw_data_maf_lower.txt"), 
    "--out", paste0(temp_directory, "/raw_data_maf_lower")
  ), ignore.stdout = TRUE, ignore.stderr = TRUE)
}

callrate_samples_geq <- dplyr::arrange(
  .data = data.table::fread(paste0(temp_directory, "/raw_data_maf_geq.imiss")), 
  F_MISS
)

prop_all <- sum(
  callrate_samples[, "F_MISS"] > 1 - params[["callrate_samples"]]
) / 
  nrow(callrate_samples)

prop_maf_geq <- sum(
  callrate_samples_geq[, "F_MISS"] > 1 - params[["callrate_samples"]]
) /
  nrow(callrate_samples_geq)

prop_diff_maf_geq <- abs((prop_maf_geq - prop_all) / prop_all) >= 0.05

if (is_maf_lower) {
  callrate_samples_lower <- dplyr::arrange(
    .data = data.table::fread(paste0(temp_directory, "/raw_data_maf_lower.imiss")), 
    F_MISS
  )
  prop_maf_lower <- sum(
    callrate_samples_lower[, "F_MISS"] > 1 - params[["callrate_samples"]]
  ) / 
    nrow(callrate_samples_lower)
  prop_diff_maf_lower <- abs((prop_maf_lower - prop_all) / prop_all) >= 0.05
} else {
  prop_diff_maf_lower <- NA
}

callrate_thresholds <- sort(unique(c(params[["callrate_samples"]], 0.90, 0.95, 0.97, 0.98, 0.99)), decreasing = TRUE)

callrate_thresholds_table <- cbind(
  callrate_thresholds,
  rowSums(sapply(callrate_samples[, "F_MISS"], ">", 1 - callrate_thresholds)),
  rowSums(sapply(callrate_samples_geq[, "F_MISS"], ">", 1 - callrate_thresholds))
) %>%
  as.data.frame() %>%
  dplyr::mutate(callrate_thresholds = scales::percent(callrate_thresholds, accuracy = 0.01)) %>%
  `colnames<-`(c(
    "Call rate threshold", 
    "Samples to exclude\n(all variants)", 
    paste0("Samples to exclude\n(variants with MAF ≥ ", params[["maf_threshold"]],")")
  ))

callrate_samples_toexclude <- callrate_samples %>%
  dplyr::filter(F_MISS > 1 - params[["callrate_samples"]]) %>%
  dplyr::mutate(Call_Rate = 1 - F_MISS) %>%
  dplyr::arrange(dplyr::desc(Call_Rate)) %>%
  dplyr::select(FID, IID, Call_Rate)
```

Call rate was computed using the `--missing` command in PLINK. 

A call rate lower than `r scales::percent(params[["callrate_samples"]], accuracy = 0.01)` was observed for **`r scales::comma(nrow(callrate_samples_toexclude))`** samples, leading to their exclusion.

```{r sample-call-rate-02, eval = !is.na(prop_diff_maf_lower) & prop_diff_maf_lower}
cat(
  "Note: the sample-based call rate was different for ", 
  "variants with MAF < ", params[["maf_threshold"]],
  " (", scales::percent(prop_maf_lower, accuracy = 0.01), ") ", 
  "compared to the call rate based on all variants ", 
  "(", scales::percent(prop_all, accuracy = 0.01), ").\n\n", 
  sep = ""
)
```

```{r sample-call-rate-03, eval = !is.na(prop_diff_maf_geq) & prop_diff_maf_geq}
cat(
  "Note: the sample-based call rate was different for ",
  "variants with MAF ≥ ", params[["maf_threshold"]], 
  " (", scales::percent(prop_maf_geq, accuracy = 0.01), ") ", 
  "compared to the call rate based on all variants ",
  "(", scales::percent(prop_all, accuracy = 0.01), ").\n\n", 
  sep = ""
)
```

```{r sample-call-rate-04}
pretty_kable(
  data = callrate_thresholds_table,
  align = "c", 
  table.attr = "style='width:65%;'",
  caption = paste(
    "Number of samples to exclude for", qdap::replace_number(length(callrate_thresholds)), "call rate thresholds."
  )
)
```

```{r sample-call-rate-05, fig.cap = "Sample call rate."}
ggplot2::ggplot(
  data = callrate_samples,
  mapping = ggplot2::aes(x = length(F_MISS):1, y = 1 - F_MISS)
) +
  ggplot2::geom_point(
    colour = scales::viridis_pal(begin = 0.5, end = 0.5)(1), 
    shape = 1, 
    na.rm = TRUE
  ) +
  ggrepel::geom_label_repel(
    mapping = ggplot2::aes(label = labs),
    min.segment.length = ggplot2::unit(0, "lines"),
    force = 10, 
    alpha = 0.7, 
    size = 2.5, 
    na.rm = TRUE
  ) +
  ggplot2::geom_hline(
    mapping = ggplot2::aes(yintercept = params[["callrate_samples"]]),
    colour = "firebrick2", 
    linetype = 2
  ) +
  ggplot2::scale_x_continuous(labels = scales::comma_format()) +
  ggplot2::scale_y_continuous(labels = scales::percent_format(accuracy = 0.01)) +
  ggplot2::labs(x = "Samples", y = "Call rate")
```

```{r sample-call-rate-06, eval = nrow(callrate_samples_toexclude) != 0}
callrate_samples_toexclude %>% 
  dplyr::mutate(Call_Rate = scales::percent(Call_Rate, accuracy = 0.01)) %>% 
  pretty_kable(
    align = "c", 
    table.attr = "style='width:65%;'",
    caption = paste0(
      "Samples to be excluded for a call rate threshold of ", 
      scales::percent(params[["callrate_samples"]], accuracy = 0.01), 
      "."
    )
  )
```


## Gender

```{r gender-check-01}
system(command = paste(
  params[["bin_path"]][["plink"]], 
  "--bfile", paste0(temp_directory, "/raw_data"), 
  "--check-sex",
  "--out", paste0(temp_directory, "/raw_data")
), ignore.stdout = TRUE, ignore.stderr = TRUE)

gender_file_check <- file.exists(paste0(temp_directory, "/raw_data.sexcheck"))

if (!gender_file_check) {
  cat("Warning: gender check was not achieved, due to no polymorphic X chromosome in data.\n\n")
}
```

```{r gender-check-02, eval = gender_file_check, include = FALSE}
sexcheck <- data.table::fread(paste0(temp_directory, "/raw_data.sexcheck")) %>%
  dplyr::mutate_at(.vars = c("FID", "IID"), as.character) %>% 
  dplyr::full_join(y = callrate_samples, by = c("FID", "IID")) %>%
  dplyr::left_join(y = recruit_centre, by = c("IID" = "Sample.Name")) %>%
  dplyr::mutate(
    sex_discrepancy = Gender != SNPSEX,
    STATUS = ifelse(sex_discrepancy | is.na(sex_discrepancy) | `F` >= 0.20 & `F` < 0.80, "PROBLEM", STATUS),
    labs = ifelse(STATUS == "PROBLEM" & F_MISS < 1 - params[["callrate_samples"]], IID, NA)
  )

sexcheck_problem <- dplyr::filter(sexcheck, STATUS == "PROBLEM")

sexcheck_discrepancy <- sexcheck_problem %>%
  dplyr::filter(sex_discrepancy) %>%
  dplyr::select(FID, IID) %>% 
  dplyr::mutate(QC = "Gender_Discrepancy")

sexcheck_missing <- sexcheck_problem %>%
  dplyr::filter(is.na(sex_discrepancy)) %>%
  dplyr::select(FID, IID) %>% 
  dplyr::mutate(QC = "Gender_Missing")

sexcheck_other <- sexcheck_problem %>%
  dplyr::filter(!sex_discrepancy) %>%
  dplyr::select(FID, IID) %>% 
  dplyr::mutate(QC = "Gender_F")

samples_toexclude <- dplyr::bind_rows(
  samples_toexclude,
  sexcheck_discrepancy,
  sexcheck_missing,
  sexcheck_other
)
```

The mean homozygosity rate across the X-chromosome was computed using the `--check-sex` command in PLINK.

For **`r scales::comma(nrow(sexcheck_problem))`** samples 
(including **`r scales::comma(sum(sexcheck_problem[, "Gender"] == 0))`**, 
with unspecified sex and **`r scales::comma(sum(sexcheck_problem[, "SNPSEX"] == 0))`**
samples with undetermined genotypic sex), the gender discrepancy between
genotyping data and reported gender could not be resolved, and have been flagged.

`r if (!file.exists(cohort_details)) cat("Note: Gender phenotype was read from PLINK files.\n")`

```{r gender-check-03, eval = gender_file_check, fig.cap = paste("The homozygosity rate for each sample (estimated using X-chromosomal variants)", "plotted against the proportion of missing genotypes per sample.", "Male samples are expected to have a homozygosity rate greater than 0.8,", "female samples are expected to have a homozygosity rate less than 0.2.", "Samples for which gender information is discordant or missing were denoted by red cross.")}
ggplot2::ggplot(
  data = tidyr::drop_na(sexcheck, `F`),
  mapping = ggplot2::aes(
    x = `F`, 
    y = 1 - F_MISS, 
    colour = STATUS, 
    shape = sex_discrepancy
  )
) +
  ggplot2::geom_rect(
    mapping = ggplot2::aes(xmin = -Inf, xmax = Inf, ymin = -Inf, ymax = params[["callrate_samples"]]), 
    fill = "grey95",
    colour = "transparent"
  ) +
  ggplot2::annotate(
    geom = "rect",
    xmin = -Inf, xmax = -0.2, ymin = -Inf, ymax = Inf, 
    fill = "firebrick2",
    alpha = 0.10,
    colour = "transparent"
  ) +
  ggplot2::annotate(
    geom = "rect",
    xmin = -0.2, xmax = 0.2, ymin = -Inf, ymax = Inf, 
    fill = "firebrick2",
    alpha = 0.2,
    colour = "transparent"
  ) +
  ggplot2::annotate(
    geom = "rect",
    xmin = 0.8, xmax = 1, ymin = -Inf, ymax = Inf, 
    fill = "dodgerblue",
    alpha = 0.2,
    colour = "transparent"
  ) +
  ggplot2::geom_point(na.rm = TRUE) +
  ggplot2::geom_hline(yintercept = params[["callrate_samples"]], colour = "firebrick2", linetype = 2) +
  ggrepel::geom_label_repel(
    mapping = ggplot2::aes(label = labs),
    min.segment.length = ggplot2::unit(0, "lines"),
    force = 10, 
    alpha = 0.7, 
    size = 2.5, 
    na.rm = TRUE
  ) +
  ggplot2::scale_colour_manual(values = c(scales::viridis_pal(begin = 0.5, end = 0.5)(1), "firebrick2")) +
  ggplot2::scale_shape_manual(values = c(1, 3)) +
  ggplot2::scale_x_continuous(breaks = seq(0, 1, 0.2)) +
  ggplot2::scale_y_continuous(labels = scales::percent_format(accuracy = 0.01)) +
  ggplot2::labs(x = "Homozygosity rate", y = "Call rate") +
  ggplot2::theme(legend.position = "none")
```


## Heterozygosity

```{r heterozygosity-01, include = FALSE}
system(command = paste(
  params[["bin_path"]][["plink"]], 
  "--bfile", paste0(temp_directory, "/raw_data"), 
  "--het",
  "--out", paste0(temp_directory, "/raw_data_auto")
), ignore.stdout = TRUE, ignore.stderr = TRUE)

if (is_maf_lower) {
  system(command = paste(
    params[["bin_path"]][["plink"]], 
    "--bfile", paste0(temp_directory, "/raw_data"),
    "--extract", paste0(temp_directory, "/raw_data_maf_lower.txt"),
    "--het",
    "--out", paste0(temp_directory, "/raw_data_maf_lower")
  ), ignore.stdout = TRUE, ignore.stderr = TRUE)
}

heterozygosity <- data.table::fread(paste0(temp_directory, "/raw_data_auto.het")) %>%
  dplyr::mutate_at(.vars = c("FID", "IID"), as.character) %>% 
  dplyr::rename(N.NM. = `N(NM)`, O.HOM. = `O(HOM)`, E.HOM. = `E(HOM)`) %>%
  dplyr::full_join(y = callrate_samples, by = c("FID", "IID")) %>%
  dplyr::group_by(F_MISS < 1 - params[["callrate_samples"]]) %>% 
  dplyr::do(
    dplyr::mutate(.,
      hrate = (N.NM. - O.HOM.) / N.NM.,
      colour = sqrt(
        (hrate - mean(hrate, na.rm = TRUE))^2 + 
          (F_MISS - mean(F_MISS, na.rm = TRUE))^2
      ), 
      labs = ifelse(
        test = F_MISS < 1 - params[["callrate_samples"]] & ( 
          hrate > mean(hrate) + stats::sd(hrate * params[["heterozygosity_treshold"]]) | 
            hrate < mean(hrate) - stats::sd(hrate * params[["heterozygosity_treshold"]])
        ), 
        yes = IID, 
        no = NA
      )
    )
  ) %>% 
  dplyr::ungroup()

if (is_maf_lower) {
  heterozygosity_maf_lower <- data.table::fread(paste0(temp_directory, "/raw_data_maf_lower.het")) %>%
    dplyr::mutate_at(.vars = c("FID", "IID"), as.character) %>% 
    dplyr::rename(N.NM. = `N(NM)`, O.HOM. = `O(HOM)`, E.HOM. = `E(HOM)`) %>%
    dplyr::full_join(y = callrate_samples, by = c("FID", "IID")) %>%
    dplyr::group_by(F_MISS < 1 - params[["callrate_samples"]]) %>% 
    dplyr::do(
      dplyr::mutate(.,
        hrate = (N.NM. - O.HOM.) / N.NM.,
        colour = sqrt((hrate - mean(hrate, na.rm = TRUE))^2 + (F_MISS - mean(F_MISS, na.rm = TRUE))^2), 
        labs = ifelse(
          test = F_MISS < 1 - params[["callrate_samples"]] & ( 
            hrate > mean(hrate) + stats::sd(hrate * params[["heterozygosity_treshold"]]) | 
              hrate < mean(hrate) - stats::sd(hrate * params[["heterozygosity_treshold"]])
          ), 
          yes = IID, 
          no = NA
        )
      )
    ) %>% 
  dplyr::ungroup()
} else {
  heterozygosity_maf_lower <- data.frame(matrix(
    NA,
    nrow = 0,
    ncol = ncol(heterozygosity),
    dimnames = list(NULL, colnames(heterozygosity))
  ))
}
```

The observed heterozygosity rate per sample was computed using the formula  
(Number of non-missing genotypes N(NM) - Number of homozygous genotypes O(Hom)) / N(NM).

N(NM) and O(Hom) were obtained using the `--het` command in PLINK.

```{r heterozygosity-02, include = FALSE}
p_heteR <- ggplot2::ggplot(
  data = heterozygosity,
  mapping = ggplot2::aes(x = 1 - F_MISS, y = hrate)
) +
  ggplot2::geom_rect(
    mapping = ggplot2::aes(ymin = -Inf, ymax = Inf, xmin = -Inf, xmax = params[["callrate_samples"]]), 
    fill = "grey95",
    colour = "transparent"
  ) +
  ggplot2::geom_point(
    mapping = ggplot2::aes(colour = colour),
    shape = 1, 
    na.rm = TRUE
  ) +
  ggplot2::geom_vline(
    mapping = ggplot2::aes(xintercept = params[["callrate_samples"]]),
    colour = "firebrick2", 
    linetype = 2
  ) +
  ggplot2::geom_hline(
    data = dplyr::filter(heterozygosity, F_MISS < 1 - params[["callrate_samples"]]),
    mapping = ggplot2::aes(
      yintercept = mean(hrate, na.rm = TRUE) - 
        stats::sd(hrate, na.rm = TRUE) * params[["heterozygosity_treshold"]]
    ),
    colour = "firebrick2", 
    linetype = 2
  ) +
  ggplot2::geom_hline(
    data = dplyr::filter(heterozygosity, F_MISS < 1 - params[["callrate_samples"]]),
    mapping = ggplot2::aes(
      yintercept = mean(hrate, na.rm = TRUE) + 
        stats::sd(hrate, na.rm = TRUE) * params[["heterozygosity_treshold"]]
    ),
    colour = "firebrick2", 
    linetype = 2
  ) +
  ggrepel::geom_label_repel(
    ggplot2::aes(label = labs),
    min.segment.length = ggplot2::unit(0, "lines"),
    force = 30, 
    alpha = 0.7, 
    size = 2.5, 
    na.rm = TRUE
  ) + 
  ggplot2::scale_colour_viridis_c(begin = 0.2, end = 0.8, trans = "sqrt") +
  ggplot2::labs(x = "Call rate", y = "Heterozygosity rate") +
  ggplot2::theme(
    axis.text.x = ggplot2::element_text(angle = 45, hjust = 1), 
    legend.position = "none"
  )

if (is_maf_lower) {
  p_heteR_maf_lower <- ggplot2::ggplot(
    data = heterozygosity_maf_lower,
    mapping = ggplot2::aes(x = 1 - F_MISS, y = hrate)
  ) +
    ggplot2::geom_rect(
      mapping = ggplot2::aes(ymin = -Inf, ymax = Inf, xmin = -Inf, xmax = params[["callrate_samples"]]), 
      fill = "grey95",
      colour = "transparent"
    ) +
    ggplot2::geom_point(
      mapping = ggplot2::aes(colour = colour),
      shape = 1, 
      na.rm = TRUE
    ) +
    ggplot2::geom_vline(
      mapping = ggplot2::aes(xintercept = params[["callrate_samples"]]),
      colour = "firebrick2", 
      linetype = 2
    ) +
    ggplot2::geom_hline(
      data = dplyr::filter(heterozygosity_maf_lower, F_MISS < 1 - params[["callrate_samples"]]),
      mapping = ggplot2::aes(
        yintercept = mean(hrate, na.rm = TRUE) - 
          stats::sd(hrate, na.rm = TRUE) * params[["heterozygosity_treshold"]]
      ),
      colour = "firebrick2", 
      linetype = 2
    ) +
    ggplot2::geom_hline(
      data = dplyr::filter(heterozygosity_maf_lower, F_MISS < 1 - params[["callrate_samples"]]),
      mapping = ggplot2::aes(
        yintercept = mean(hrate, na.rm = TRUE) + 
          stats::sd(hrate, na.rm = TRUE) * params[["heterozygosity_treshold"]]
      ),
      colour = "firebrick2", 
      linetype = 2
    ) +
    ggrepel::geom_label_repel(
      ggplot2::aes(label = labs),
      min.segment.length = ggplot2::unit(0, "lines"),
      force = 30, 
      alpha = 0.7, 
      size = 2.5, 
      na.rm = TRUE
    )+
    ggplot2::scale_colour_viridis_c(begin = 0.2, end = 0.8, trans = "sqrt") +
    ggplot2::labs(x = "Call rate", y = "Heterozygosity rate") +
    ggplot2::theme(
      axis.text.x = ggplot2::element_text(angle = 45, hjust = 1), 
      legend.position = "none"
    ) +
    ggplot2::labs(
      x = paste0("Call rate (MAF < ", params[["maf_threshold"]], ")"), 
      y = "Heterozygosity rate"
    ) +
    ggplot2::theme(
      axis.text.x = ggplot2::element_text(angle = 45, hjust = 1), 
      legend.position = "none"
    )
} else {
  p_heteR_maf_lower <- ggplot2::ggplot(data = heterozygosity_maf_lower) +
    ggplot2::geom_blank() +
    ggplot2::theme(
      axis.line = ggplot2::element_blank(),
      axis.text = ggplot2::element_blank(),
      axis.ticks = ggplot2::element_blank(),
      axis.title = ggplot2::element_blank(),
      legend.position = "none",
      panel.background = ggplot2::element_blank(),
      panel.border = ggplot2::element_blank(),
      panel.grid = ggplot2::element_blank(),
      plot.background = ggplot2::element_blank()
    ) +
    ggplot2::annotate(
      "text",
      label = paste0("No variants found with MAF < ", params[["maf_threshold"]]),
      x = mean(stats::na.omit(heterozygosity[["F_MISS"]])),
      y = mean(stats::na.omit(heterozygosity[["hrate"]]))
    )
}

samples_toexclude <- dplyr::bind_rows(
  samples_toexclude,
  heterozygosity %>%
    dplyr::filter(!is.na(labs)) %>%
    dplyr::select(FID, IID) %>% 
    dplyr::mutate(QC = "Heterozygosity_Check")
)

if (is_maf_lower) {
  samples_toexclude <- dplyr::bind_rows(
    samples_toexclude,
    heterozygosity_maf_lower %>%
      dplyr::filter(!is.na(labs)) %>%
      dplyr::select(FID, IID) %>% 
      dplyr::mutate(QC = paste0("Heterozygosity_Check_maf_", params[["maf_threshold"]]))
  )
}
```

```{r heterozygosity-05, fig.height = 4.7 * 1.5, fig.cap = paste0("Heterozygosity rate against the call rate per sample. ", "A and C, the heterozygosity rate computed based on all variants. ", "B and D, the heterozygosity estimated from variants with a MAF < ", params[["maf_threshold"]], ". ", "The horizontal red lines denote the ", qdap::replace_number(params[["heterozygosity_treshold"]]), " times standard deviations (SD) from the mean heterozygosity rate ", "and were used to define outlying heterozygosity.", " SD was computed only on samples with a call rate greater than or equal to ", scales::percent(params[["callrate_samples"]], accuracy = 0.01), ".")}
ggpubr::ggarrange(
  p_heteR + 
    ggplot2::scale_x_continuous(labels = scales::percent_format(accuracy = 0.01)) +
    ggplot2::scale_y_continuous(labels = scales::percent_format(accuracy = 0.01)), 
  p_heteR_maf_lower + 
    ggplot2::scale_x_continuous(labels = scales::percent_format(accuracy = 0.01)) +
    ggplot2::scale_y_continuous(labels = scales::percent_format(accuracy = 0.01)), 
  p_heteR + 
    ggplot2::scale_x_continuous(
      labels = scales::percent_format(accuracy = 0.01),
      limits = c(params[["callrate_samples"]], NA)
    ) +
    ggplot2::scale_y_continuous(
      labels = scales::percent_format(accuracy = 0.01),
      limits = range(
        c(
          dplyr::filter(heterozygosity, F_MISS < 1 - params[["callrate_samples"]])[["hrate"]],
          dplyr::filter(heterozygosity, F_MISS < 1 - params[["callrate_samples"]])[["hrate"]] %>% 
            (function(x) {
              c(
                mean(x, na.rm = TRUE) + 
                  stats::sd(x, na.rm = TRUE) * params[["heterozygosity_treshold"]],
                mean(x, na.rm = TRUE) - 
                  stats::sd(x, na.rm = TRUE) * params[["heterozygosity_treshold"]]
              )
            })(.)
        )
      )
    ),
  p_heteR_maf_lower + 
    ggplot2::scale_x_continuous(
      labels = scales::percent_format(accuracy = 0.01),
      limits = c(params[["callrate_samples"]], NA)
    ) +
    ggplot2::scale_y_continuous(
      labels = scales::percent_format(accuracy = 0.01),
      limits = range(
        c(
          dplyr::filter(heterozygosity_maf_lower, F_MISS < 1 - params[["callrate_samples"]])[["hrate"]],
          dplyr::filter(heterozygosity_maf_lower, F_MISS < 1 - params[["callrate_samples"]])[["hrate"]] %>% 
            (function(x) {
              c(
                mean(x, na.rm = TRUE) + 
                  stats::sd(x, na.rm = TRUE) * params[["heterozygosity_treshold"]],
                mean(x, na.rm = TRUE) - 
                  stats::sd(x, na.rm = TRUE) * params[["heterozygosity_treshold"]]
              )
            })(.)
        )
      )
    ),
  ncol = 2, 
  nrow = 2, 
  labels = LETTERS
)
```


## Relatedness

`r if (params[["includes_relatives"]]) cat("Not applicable.\n\n")`

```{r relatedness-01, eval = !params[["includes_relatives"]], include = FALSE}
## Pruning Data
system(command = paste(
  params[["bin_path"]][["plink"]], 
  "--bfile", paste0(temp_directory, "/raw_data"), 
  "--autosome", 
  "--set-hh-missing", 
  "--maf", params[["maf_threshold"]], 
  "--r2", 
  "--ld-window-r2 0.2", 
  "--ld-window-kb 50", 
  "--make-bed",
  "--out", paste0(temp_directory, "/raw_data_pruned")
), ignore.stdout = TRUE, ignore.stderr = TRUE)

raw_data_pruned <- data.table::fread(paste0(temp_directory, "/raw_data_pruned.fam"))
max_n_pairs <- nrow(raw_data_pruned) * (nrow(raw_data_pruned) - 1 ) / 2

system(command = paste(
  params[["bin_path"]][["plink"]], 
  "--bfile", paste0(temp_directory, "/raw_data_pruned"), 
  "--freq", 
  "--out ", paste0(temp_directory, "/raw_data_pruned")
), ignore.stdout = TRUE, ignore.stderr = TRUE)

system(command = paste(
  params[["bin_path"]][["plink"]], 
  "--bfile", paste0(temp_directory, "/raw_data_pruned"), 
  "--read-freq", paste0(temp_directory, "/raw_data_pruned.frq"), 
  "--genome", 
  "--min", params[["IBD_threshold"]],  
  "--out", paste0(temp_directory, "/raw_data_pruned_Fast")
), ignore.stdout = TRUE, ignore.stderr = TRUE)

relatedness <- data.table::fread(paste0(temp_directory, "/raw_data_pruned_Fast.genome")) %>%
  dplyr::distinct() %>%
  dplyr::mutate(PI_HAT = as.numeric(PI_HAT)) %>% 
  dplyr::mutate_at(.vars = dplyr::vars(dplyr::matches("FID|IID")), .funs = as.character) %>% 
  dplyr::anti_join(
    y = samples_toexclude, 
    by = c("FID1" = "FID", "IID1" = "IID")
  ) %>% 
  dplyr::anti_join(
    y = samples_toexclude, 
    by = c("FID2" = "FID", "IID2" = "IID")
  )
```

In order to identify pairs of samples (N = `r scales::comma(nrow(raw_data_pruned))`, 
Pairs = `r scales::comma(max_n_pairs)`) who look too similar to each other, 
pairwise IBD (identity by descent) was computed, using the `--genome` command in PLINK.

Prior to the analysis, the dataset was pruned to exclude high-LD-regions, 
that is no variants with MAF < `r params[["maf_threshold"]]` and no pair of variants within 50 Kb have an r² > 0.2.

If samples in the cohort **`r params[["cohort_name"]]`** are not supposed to be related, 
the expected pairwise $\hat{\pi}$ ("PI_HAT") should be less than `r params[["IBD_threshold"]]`.  
Pairs of sample with $\hat{\pi}$ ("PI_HAT") greater than `r params[["IBD_threshold"]]` were flagged 
(Table \@ref(tab:relatedness-03)).

```{r relatedness-02, eval = !params[["includes_relatives"]]}
data.frame(table(cut(
  relatedness[, "PI_HAT"],
  breaks = c(params[["IBD_threshold"]], 0.3, 0.4, 0.6, 0.8, 1),
  include.lowest = TRUE
))) %>%
  `colnames<-`(c("IBD windows", "Pair count")) %>%
  dplyr::mutate(
    `IBD windows` = paste0(
      `IBD windows`,
      c(" = Second degree relatives", "", " = First degree relatives", "", " = MZ twins/duplicates")
    )
  ) %>% 
  pretty_kable(
    align = "c", 
    table.attr = "style='width:65%;'",
    caption = paste0(
      'The distribution of "PI_HAT" (proportion of IBD) for related samples in the dataset ',
      "(IBD ≥ ", params[["IBD_threshold"]], ")."
    )
  )
```

```{r relatedness-03, eval = !params[["includes_relatives"]]}
related_samples <- relatedness %>%
  dplyr::select(FID1, IID1, FID2, IID2, PI_HAT) %>%
  dplyr::left_join(
    callrate_samples %>%
      dplyr::select(FID, IID, F_MISS) %>%
      dplyr::mutate_at(.vars = 1:2, as.character),
    by = c("FID1" = "FID", "IID1" = "IID")
  ) %>%
  dplyr::left_join(
    callrate_samples %>%
      dplyr::select(FID, IID, F_MISS) %>%
      dplyr::mutate_at(.vars = 1:2, as.character),
    by = c("FID2" = "FID", "IID2" = "IID"),
    suffix = c("1", "2")
  ) %>%
  dplyr::mutate(
    Call_Rate1 = scales::percent(1 - F_MISS1, accuracy = 0.01),
    Call_Rate2 = scales::percent(1 - F_MISS2, accuracy = 0.01)
  ) %>%
  dplyr::select(-F_MISS1, -F_MISS2) %>% 
  dplyr::arrange(dplyr::desc(PI_HAT))
                 
if (nrow(related_samples) != 0) {
    pretty_kable(
      data = related_samples,
      align = "c", 
      table.attr = "style='width:65%;'",
      caption = paste0(
        "Pairs of samples with a proportion of IBD higher than ",  params[["IBD_threshold"]], "."
      )
    )
}

if (dplyr::n_distinct(dplyr::select(recruit_centre, Cohort)) > 1) {
  cross_cohorts_IBD <- relatedness %>%
    dplyr::select(FID1, IID1, FID2, IID2, PI_HAT) %>%
    dplyr::left_join(
      y = dplyr::select(recruit_centre, -Gender),
      by = c("IID1" = "Sample.Name")
    ) %>%
    dplyr::left_join(
      y = dplyr::select(recruit_centre, -Gender),
      by = c("IID2" = "Sample.Name"),
      suffix = c("1", "2")
    ) %>%
    dplyr::filter(Cohort1 != Cohort2) %>%
    dplyr::arrange(dplyr::desc(PI_HAT))
  
  if (nrow(cross_cohorts_IBD) > 0) {
    cat(
      Hmisc::capitalize(qdap::replace_number(nrow(cross_cohorts_IBD))), 
      "cross-cohort IBDs was found in the dataset.\n\n"
    )
    
    pretty_kable(
      data = cross_cohorts_IBD, 
      align = "c", 
      table.attr = "style='width:65%;'",
      caption = "Cross cohort related sample pairs."
    )
  } else {
    cat("No cross-cohort IBDs was found in the dataset.\n\n")
  }
}

samples_toexclude <- dplyr::bind_rows(
  samples_toexclude,
  relatedness %>%
    dplyr::filter(PI_HAT >= params[["IBD_threshold"]]) %>% 
    dplyr::select(dplyr::matches("FID|IID")) %>% 
    tidyr::unite(col = FID_IID1, FID1, IID1, sep = "_-_") %>% 
    tidyr::unite(col = FID_IID2, FID2, IID2, sep = "_-_") %>% 
    tidyr::gather(key = "FID_IID", value = "value") %>% 
    tidyr::separate(col = "value", into = c("FID", "IID"), sep = "_-_") %>% 
    dplyr::select(-FID_IID) %>% 
    dplyr::distinct(FID, IID) %>% 
    dplyr::mutate(QC = "Relatedness")
)
```


## Sample Mendel errors

```{r imendel-error-00}
if (!params[["includes_relatives"]]) cat("Not applicable.\n\n")

imendel_file_check <- params[["includes_relatives"]]
```

```{r imendel-error-01, eval = params[["includes_relatives"]]}
system(command = paste(
  params[["bin_path"]][["plink"]], 
  "--bfile", paste0(temp_directory, "/raw_data"), 
  "--mendel",
  "--out ", paste0(temp_directory, "/raw_data") 
), ignore.stdout = TRUE, ignore.stderr = TRUE)

system(command = paste(
  params[["bin_path"]][["plink"]], 
  "--bfile", paste0(temp_directory, "/raw_data"), 
  "--extract", paste0(temp_directory, "/raw_data_maf_geq.txt"), 
  "--mendel", 
  "--out", paste0(temp_directory, "/raw_data_maf_geq")
), ignore.stdout = TRUE, ignore.stderr = TRUE)

if (is_maf_lower) {
  system(command = paste(
    params[["bin_path"]][["plink"]], 
    "--bfile", paste0(temp_directory, "/raw_data"), 
    "--extract", paste0(temp_directory, "/raw_data_maf_lower.txt"), 
    "--mendel", 
    "--out", paste0(temp_directory, "/raw_data_maf_lower")
  ), ignore.stdout = TRUE, ignore.stderr = TRUE)
}

imendel_file_check <- file.exists(paste0(temp_directory, "/raw_data.imendel"))

if (!imendel_file_check) {
  cat("Mendel error rates check was not performed due to missing informations.\n\n")
}
```

```{r imendel-error-02, eval = params[["includes_relatives"]] & imendel_file_check}
mendel_errors <- data.table::fread(paste0(temp_directory, "/raw_data.imendel")) %>%
  dplyr::mutate_at(.vars = dplyr::vars(FID, IID), as.character) %>% 
  dplyr::mutate(
    percentage = N / snps_number,
    labs = ifelse(percentage > params[["mendelian_samples"]], IID, NA),
    labs_fam = ifelse(percentage > params[["mendelian_samples"]], FID, NA)
  )

# MAF >= 0.01
mendel_errors_geq <- data.table::fread(paste0(temp_directory, "/raw_data_maf_geq.imendel")) %>%
  dplyr::mutate(percentage = N / snps_number)

prop_maf_geq <- sum(mendel_errors_geq[["percentage"]] > params[["mendelian_samples"]]) / 
  nrow(mendel_errors_geq)

prop_all <- sum(mendel_errors[["percentage"]] > params[["mendelian_samples"]]) / 
  nrow(mendel_errors)

prop_diff_maf_geq <- abs((prop_maf_geq - prop_all) / prop_all) >= 0.05

# MAF < 0.01
if (is_maf_lower) {
  mendel_errors_lower <- data.table::fread(paste0(temp_directory, "/raw_data_maf_lower.imendel")) %>% 
    dplyr::mutate(percentage = N / snps_number)
  
  prop_maf_lower <- sum(mendel_errors_lower[["percentage"]] > params[["mendelian_samples"]]) /
    nrow(mendel_errors_lower)
  
  prop_diff_maf_lower <- abs((prop_maf_lower - prop_all) / prop_all) >= 0.05
}

cat(
  "The sample-wise Mendel error was estimated using the `--mendel` command in PLINK.  ", "\n",
  "The per-trio error rate threshold is ", params[["mendelian_samples"]], ".",
  "\n\n",
  "Figure \\@ref(fig:imendel-error-03) shows the Mendel error per sample.  ", "\n",
  scales::comma(nrow(mendel_errors)), " samples (from ", dplyr::n_distinct(mendel_errors[["FID"]]), " families)",
  " had a Mendel error rate higher than ", scales::percent(x = params[["mendelian_samples"]], accuracy = 0.01), 
  " and were flagged (Table \\@ref(tab:imendel-error-04)).",
  "\n\n",
  sep = ""
)

if (!is.na(prop_diff_maf_lower) & prop_diff_maf_lower) {
  cat(
    "Sample-wise Mendel error rate ratio is different for variants ",
      "with MAF < 0.01 (", scales::percent(prop_maf_lower, accuracy = 0.01), ") ",
      "compared to Mendel error rate ratio for all variants ",
      "(", scales::percent(prop_all, accuracy = 0.01), ").\n\n",
    sep = ""
  )
}

if (!is.na(prop_diff_maf_geq) & prop_diff_maf_geq) {
  cat(
    "Sample-wise Mendel error rate ratio is different for variants ",
      "with MAF > 0.01 (", scales::percent(prop_maf_geq),") ",
      "compared to Mendel error rate ratio for all variants ",
      "(", scales::percent(prop_all, accuracy = 0.01), ").\n\n",
    sep = ""
  )
}

mendel_threshold <- sort(unique(c(params[["mendelian_samples"]], 0.05, 0.03, 0.02, 0.01)), decreasing = TRUE)

cbind(
  mendel_threshold,
  rowSums(sapply(mendel_errors[["percentage"]], ">", mendel_threshold)),
  rowSums(sapply(mendel_errors_geq[["percentage"]], ">", mendel_threshold))
) %>%
  as.data.frame() %>%
  dplyr::mutate(mendel_threshold = scales::percent(x = mendel_threshold, accuracy = 0.01)) %>%
  `colnames<-`(c("Sample Mendel error threshold", "Samples to exclude", "Samples to exclude (MAF > 0.01)")) %>% 
  pretty_kable(
    align = "c", 
    table.attr = "style='width:65%;'",
    caption = paste(
      "Number of samples to exclude for", qdap::replace_number(length(mendel_threshold)), "call rate thresholds."
    )
  )
```

```{r imendel-error-03, eval = params[["includes_relatives"]] & imendel_file_check, fig.cap = "Sample Mendel error rate."}
ggplot2::ggplot(
  data = dplyr::arrange(mendel_errors, percent),
  mapping = ggplot2::aes(
    x = seq_len(nrow(mendel_errors)), 
    y = percent
  )
) +
  ggplot2::geom_point(
    colour = scales::viridis_pal(begin = 0.5, end = 0.5)(1), 
    shape = 1, 
    na.rm = TRUE
  ) +
  ggplot2::geom_hline(
    mapping = ggplot2::aes(
      yintercept = params[["mendelian_samples"]]
    ), 
    colour = "firebrick2", 
    linetype = 2
  ) +
  ggrepel::geom_label_repel(
    mapping = ggplot2::aes(label = labs),
    min.segment.length = ggplot2::unit(0, "lines"),
    segment.colour = "black", 
    colour = "black",
    force = 10, 
    alpha = 0.7, 
    size = 2.5,
    na.rm = TRUE
  ) +
  ggplot2::scale_x_continuous(labels = scales::comma_format()) +
  ggplot2::scale_y_continuous(labels = scales::percent_format(accuracy = 0.01)) +
  ggplot2::labs(x = "Samples", y = "Mendel Error Rate")
```

```{r imendel-error-04, eval = params[["includes_relatives"]] & imendel_file_check}
mendel_errors_samples <- mendel_errors %>%
  dplyr::filter(percentage > params[["mendelian_samples"]]) %>%
  dplyr::select(FID, IID, percentage) %>%
  dplyr::arrange(FID, IID) %>% 
  dplyr::mutate(QC = "Sample_Mendel_Error")

if (nrow(mendel_errors_samples)>0) {
  mendel_errors_samples %>% 
    dplyr::mutate(
      `Sample Mendel error rate` = scales::percent(x = percent, accuracy = 0.01)
    ) %>% 
    dplyr::select(-percent, -QC) %>% 
    pretty_kable(
      align = "c", 
      table.attr = "style='width:65%;'",
      caption = paste0(
        "Samples to be checked for max per-trio error rate at ", 
        scales::percent(x = params[["mendelian_samples"]], accuracy = 0.01), "."
      )
    )
}

samples_toexclude <- dplyr::bind_rows(
  samples_toexclude, 
  mendel_errors_samples
)
```


## Principal Component Analysis


### Population structure

```{r population-01, include = FALSE}
## Get Autosomal variants
data.table::fread(paste0(temp_directory, "/raw_data.bim")) %>%
  dplyr::filter(V1 %in% seq(22)) %>%
  dplyr::select(V2) %>%
  utils::write.table(
    file = paste0(temp_directory, "/raw_data_auto.txt"),
    quote = FALSE, 
    col.names = FALSE, 
    row.names = FALSE
  )

## Remove bad quality samples (Call Rate and Heterozygosity)
samples_toexclude_pca <- samples_toexclude %>% 
  dplyr::filter(grepl("Sample_Call_Rate|Heterozygosity_Check$", QC)) %>% 
  dplyr::distinct(FID, IID) %>% 
  dplyr::select(FID, IID)

utils::write.table(
  x = samples_toexclude_pca,
  file = paste0(temp_directory, "/samples_toexclude_pca.txt"),
  sep = "\t", 
  row.names = FALSE, 
  col.names = FALSE, 
  quote = FALSE
)

if (nrow(samples_toexclude_pca) > 0) {
  system(command = paste(
    params[["bin_path"]][["plink"]], 
    "--bfile", paste0(temp_directory, "/raw_data"),
    "--remove", paste0(temp_directory, "/samples_toexclude_pca.txt"),
    "--extract", paste0(temp_directory, "/raw_data_auto.txt"),
    "--snps-only",
    "--maf", params[["maf_threshold"]], 
    "--hwe", params[["hwe_pvalue"]], 
    "--geno 0.1", 
    "--make-bed",
    "--out", paste0(temp_directory, "/clean_data_tmp")
  ), ignore.stdout = TRUE, ignore.stderr = TRUE)
} else {
  system(command = paste(
    params[["bin_path"]][["plink"]], 
    "--bfile", paste0(temp_directory, "/raw_data"), 
    "--extract", paste0(temp_directory, "/raw_data_auto.txt"),
    "--snps-only",
    "--maf", params[["maf_threshold"]], 
    "--hwe", params[["hwe_pvalue"]], 
    "--geno 0.1", 
    "--make-bed",
    "--out", paste0(temp_directory, "/clean_data_tmp")
  ), ignore.stdout = TRUE, ignore.stderr = TRUE)
}

## Extract 1kgenome data
data.table::fread(paste0(temp_directory, "/clean_data_tmp.bim")) %>%
  dplyr::filter(!V4 %in% unique(V4[duplicated(V4)])) %>%
  dplyr::select(V2) %>%
  utils::write.table(
    file = paste0(temp_directory, "/clean_data_snp.txt"),
    row.names = FALSE, 
    col.names = FALSE, 
    quote = FALSE
  )

parallel::mclapply(seq_len(22), mc.cores = min(params[["n_cores"]], 22), function(i) {
  system(command = paste(
    params[["bin_path"]][["plink"]], 
    "--bfile", paste0(params[["ref1kg_genotypes"]], "Chr", i),
    "--extract", paste0(temp_directory, "/clean_data_snp.txt"), 
    "--snps-only",
    "--make-bed", 
    "--out", paste0(temp_directory, "/1kgenome_small/1kgenome_small_chr", sprintf("%02d", i))
  ), ignore.stdout = TRUE, ignore.stderr = TRUE)
})

raw_data_1kgenome_files <- t(sapply(
  X = paste0(temp_directory, "/1kgenome_small/1kgenome_small_chr", sprintf("%02d", seq_len(22))),
  FUN = paste0, c(".bed", ".bim", ".fam"),
  simplify = TRUE, 
  USE.NAMES = FALSE
))

list_1kgenome <- paste0(temp_directory, "/1kgenome_small/1kgenome_small_files.txt")

utils::write.table(
  x = raw_data_1kgenome_files,
  file = list_1kgenome,
  col.names = FALSE, 
  row.names = FALSE, 
  quote = FALSE
)

first_chromosome <- gsub(".bed", "", raw_data_1kgenome_files[1, 1])

out <- paste0(temp_directory, "/1kgenome_small/1kgenome_small_complete")

system(command = paste(
  params[["bin_path"]][["plink"]], 
  "--bfile", first_chromosome, 
  "--merge-list", list_1kgenome, 
  "--snps-only", 
  "--make-bed", 
  "--out", out
), ignore.stdout = TRUE, ignore.stderr = TRUE)

if (file.exists(paste0(out, "-merge.missnp"))) {
  parallel::mclapply(seq_len(22), mc.cores = 22, function(i) {
    system(command = paste(
      params[["bin_path"]][["plink"]], 
      "--bfile", paste0(params[["ref1kg_genotypes"]], "Chr", i),
      "--exclude", paste0(out, "-merge.missnp"),
      "--extract", paste0(temp_directory, "/clean_data_snp.txt"), 
      "--snps-only",
      "--make-bed",
      "--out", paste0(temp_directory, "/1kgenome_small/1kgenome_small_chr", sprintf("%02d", i))
    ), ignore.stdout = TRUE, ignore.stderr = TRUE)
  })
  system(command = paste(
    params[["bin_path"]][["plink"]], 
    "--bfile", first_chromosome, 
    "--exclude", paste0(out, "-merge.missnp"), 
    "--merge-list", list_1kgenome, 
    "--snps-only",
    "--make-bed", 
    "--out", out
  ), ignore.stdout = TRUE, ignore.stderr = TRUE)
}

## Merge raw_data with 1kgenome
# Merge raw_data with 1kGenome; ERROR: Stopping due to mis-matching variants -- check +/- strand?
system(command = paste(
  params[["bin_path"]][["plink"]], 
  "--bfile", paste0(temp_directory, "/1kgenome_small/1kgenome_small_complete"),
  "--bmerge", 
    paste0(temp_directory, "/clean_data_tmp.bed"), 
    paste0(temp_directory, "/clean_data_tmp.bim"),
    paste0(temp_directory, "/clean_data_tmp.fam"), 
  "--make-bed", 
  "--out", paste0(temp_directory, "/raw_data_1kgenome")
), ignore.stdout = TRUE, ignore.stderr = TRUE)

# Flip variant strand
system(command = paste(
  params[["bin_path"]][["plink"]], 
  "--bfile", paste0(temp_directory, "/1kgenome_small/1kgenome_small_complete"),
  "--flip", paste0(temp_directory, "/raw_data_1kgenome-merge.missnp"), 
  "--make-bed",
  "--out", paste0(temp_directory, "/1kgenome_small/1kgenome_small_flipped")
), ignore.stdout = TRUE, ignore.stderr = TRUE)

# Merge raw_data with 1kGenome; ERROR: Stopping due to mis-matching variants -- check +/- strand?
system(command = paste(
  params[["bin_path"]][["plink"]], 
  "--bfile", paste0(temp_directory, "/1kgenome_small/1kgenome_small_flipped"),
  "--bmerge", 
    paste0(temp_directory, "/clean_data_tmp.bed"), 
    paste0(temp_directory, "/clean_data_tmp.bim"),
    paste0(temp_directory, "/clean_data_tmp.fam"), 
  "--make-bed", 
  "--out", paste0(temp_directory, "/raw_data_1kgenome")
), ignore.stdout = TRUE, ignore.stderr = TRUE)

# Exclude mis-matching variants in raw_data And 1kGenome
system(command = paste(
  params[["bin_path"]][["plink"]], 
  "--bfile", paste0(temp_directory, "/1kgenome_small/1kgenome_small_flipped"),
  "--exclude", paste0(temp_directory, "/raw_data_1kgenome-merge.missnp"), 
  "--make-bed",
  "--out", paste0(temp_directory, "/1kgenome_small/1kgenome_small_flipped_cleaned")
), ignore.stdout = TRUE, ignore.stderr = TRUE)

system(command = paste(
  params[["bin_path"]][["plink"]], 
  "--bfile", paste0(temp_directory, "/clean_data_tmp"),
  "--exclude", paste0(temp_directory, "/raw_data_1kgenome-merge.missnp"), 
  "--make-bed",
  "--out", paste0(temp_directory, "/clean_data_tmp_cleaned")
), ignore.stdout = TRUE, ignore.stderr = TRUE)

# Final, Merge raw_data with 1kGenome
system(command = paste(
  params[["bin_path"]][["plink"]], 
  "--bfile", paste0(temp_directory, "/1kgenome_small/1kgenome_small_flipped_cleaned"),
  "--bmerge ", 
    paste0(temp_directory, "/clean_data_tmp_cleaned.bed"), 
    paste0(temp_directory, "/clean_data_tmp_cleaned.bim"),
    paste0(temp_directory, "/clean_data_tmp_cleaned.fam"), 
  "--make-bed", 
  "--out", paste0(temp_directory, "/raw_data_1kgenome")
), ignore.stdout = TRUE, ignore.stderr = TRUE)

# Keep only variants in common between raw_data and 1kGenome
dplyr::intersect(
 data.table::fread(paste0(temp_directory, "/1kgenome_small/1kgenome_small_flipped_cleaned.bim"))[["V2"]],
 data.table::fread(paste0(temp_directory, "/clean_data_tmp_cleaned.bim"))[["V2"]]
) %>%
  utils::write.table(
    file = paste0(temp_directory, "/raw_data_1kgenome_snps.txt"),
    col.names = FALSE, 
    row.names = FALSE, 
    quote = FALSE, 
    sep = "\n"
  )

system(command = paste(
  params[["bin_path"]][["plink"]], 
  "--bfile", paste0(temp_directory, "/raw_data_1kgenome"), 
  "--extract", paste0(temp_directory, "/raw_data_1kgenome_snps.txt"), 
  "--make-bed ",
  "--out", paste0(temp_directory, "/raw_data_1kgenome")
), ignore.stdout = TRUE, ignore.stderr = TRUE)

## Check palindromic variants
data.table::fread(paste0(temp_directory, "/raw_data_1kgenome.bim")) %>%
  tidyr::unite(alleles, V5, V6, remove = FALSE, sep = "") %>%
  dplyr::filter(alleles %in% c("GC", "CG", "AT", "TA")) %>%
  dplyr::select(V2) %>%
  utils::write.table(
    file = paste0(temp_directory, "/raw_data_1kgenome_palindromic.txt"),
    row.names = FALSE, 
    col.names = FALSE, 
    quote = FALSE
  )

system(command = paste(
  params[["bin_path"]][["plink"]], 
  "--bfile", paste0(temp_directory, "/raw_data_1kgenome"), 
  "--exclude", paste0(temp_directory, "/raw_data_1kgenome_palindromic.txt"), 
  "--make-bed ",
  "--out", paste0(temp_directory, "/raw_data_1kgenome")
), ignore.stdout = TRUE, ignore.stderr = TRUE)

## Pop studied and 1kgenome
system(command = paste(
  params[["bin_path"]][["gcta"]], 
  "--bfile", paste0(temp_directory, "/raw_data_1kgenome"),
  "--make-grm-bin", 
  "--pca", params[["pca_components"]], 
  "--thread-num", params[["n_cores"]],
  "--out", paste0(temp_directory, "/raw_data_1kgenome")
), ignore.stdout = TRUE, ignore.stderr = TRUE)

system(command = paste(
  params[["bin_path"]][["gcta"]], 
  "--grm", paste0(temp_directory, "/raw_data_1kgenome"), 
  "--pca", params[["pca_components"]], 
  "--thread-num", params[["n_cores"]], 
  "--out", paste0(temp_directory, "/raw_data_1kgenome")
), ignore.stdout = TRUE, ignore.stderr = TRUE)

## Analysis summary
# variant
n_good_snp <- nrow(data.table::fread(
  input = paste0(temp_directory, "/clean_data_tmp.bim"), 
  select = 1
))

n_good_snp_1kg <- nrow(data.table::fread(
  input = paste0(temp_directory, "/raw_data_1kgenome.bim"), 
  select = 1
))

# Samples
n_samples <- nrow(data.table::fread(
  input = paste0(temp_directory, "/raw_data_1kgenome.fam"), 
  select = 1
))

n_samples_1kg <- nrow(data.table::fread(
  input = paste0(temp_directory, "/1kgenome_small/1kgenome_small_flipped_cleaned.fam"), 
  select = 1
))
  
data_pca <- dplyr::bind_rows(
  data.table::fread(paste0(temp_directory, "/clean_data_tmp.fam")) %>%
    dplyr::mutate(POP = "raw_data") %>%
    dplyr::select(V2, POP) %>%
    dplyr::rename(IID = "V2") %>% 
    dplyr::mutate(IID = as.character(IID)),
  data.table::fread(params[["ref1kg_panel"]], fill = TRUE, header = TRUE) %>%
    dplyr::select(sample, pop) %>%
    dplyr::rename(IID = "sample", POP = "pop") %>% 
    dplyr::mutate(IID = as.character(IID))
) %>%
  dplyr::filter(
    IID %in% as.character(data.table::fread(paste0(temp_directory, "/raw_data_1kgenome.fam"))[["V2"]])
  ) %>%
  dplyr::left_join(
    y = data.table::fread(paste0(temp_directory, "/raw_data_1kgenome.eigenvec")) %>%
      `colnames<-`(c("FID", "IID", paste0("PC", seq_len(params[["pca_components"]])))) %>% 
      dplyr::mutate(IID = as.character(IID)),
    by = "IID"
  ) %>%
  dplyr::mutate(
    POP = factor(
      x = POP, 
      levels = c("raw_data", sort(setdiff(unique(POP), "raw_data"), decreasing = TRUE))
    )
  ) %>%
  dplyr::rename(x = PC1, y = PC2, Population = POP) %>%
  dplyr::select(IID, Population, x, y) %>%
  dplyr::left_join(
    y = stats::aggregate(cbind(x, y) ~ Population, data = ., mean) %>%
      `colnames<-`(paste0(colnames(.), c("", ".centroid", ".centroid"))),
    by = "Population"
  ) %>%
  dplyr::mutate(dist = sqrt((x - x.centroid)^2 + (y - y.centroid)^2)) %>%
  dplyr::group_by(Population) %>%
  dplyr::mutate(
    close = dist > (stats::quantile(dist, 0.75) + diff(stats::quantile(dist, c(0.25, 0.75))))
  ) %>%
  dplyr::ungroup()

## Compute components inertia
inertia_pca <- data.table::fread(paste0(temp_directory, "/raw_data_1kgenome.eigenval")) %>%
  tibble::rownames_to_column("x") %>%
  dplyr::mutate(y = V1 / sum(V1))
```

The data from the **`r params[["cohort_name"]]`** cohort was pruned using the following cut-offs:

* variants call rate > 90.00 %
* MAF ≥ `r params[["maf_threshold"]]`
* HWE > `r scales::scientific(params[["hwe_pvalue"]])`
* samples call rate > `r scales::percent(params[["callrate_samples"]], accuracy = 0.01)`
* heterozygosity rate within `r qdap::replace_number(params[["heterozygosity_treshold"]])` times SD from the mean heterozygosity rate
    
The resulting "clean" dataset (*i.e.*, **`r scales::comma(n_samples-n_samples_1kg)`** samples and **`r scales::comma(n_good_snp)`** variants) was merged with the data from the publicly available *1,000 Genomes project* database, 
and thinned down to only include not palindromic variants that were present in both dataset. 

A principal component analysis (PCA) was performed on the merged dataset 
(*i.e.*, **`r scales::comma(n_samples)`** samples and **`r scales::comma(n_good_snp_1kg)`** variants, 
including **`r scales::comma(n_samples_1kg)`** from the *1,000 Genomes project*)

```{r population-02, include = FALSE}
centre_pca <- data_pca %>%
  dplyr::filter(Population != "raw_data") %>%
  dplyr::select(IID, Population, dplyr::contains("centroid")) %>%
  dplyr::mutate_at(.vars = dplyr::vars(IID, Population), .funs = as.character)

check_ethnicity <- data.table::fread(paste0(temp_directory, "/clean_data_tmp.fam")) %>%
  dplyr::select(FID = 1, IID = 2) %>%
  dplyr::mutate_all(as.character) %>%
  dplyr::left_join(
    dplyr::filter(data_pca, Population == "raw_data"),
    by = "IID"
  ) %>%
  dplyr::rowwise() %>%
  dplyr::mutate(
    ClosestPop = centre_pca[["Population"]][which.min(sqrt(
      (x - centre_pca[["x.centroid"]])^2 + (y - centre_pca[["y.centroid"]])^2
    ))]
  ) %>% 
  dplyr::ungroup()

pop1kg_description <- data.table::fread(params[["ref1kg_population"]]) %>%
  dplyr::select(POP = 1, SUPER_POP = 4, SUPER_POP_NAME = 5)

if (is.null(params[["population"]])) {
  not_target_population <- check_ethnicity[0, ]
} else {
  not_target_population <- check_ethnicity %>%
    dplyr::filter(!ClosestPop %in% c(
      params[["population"]],
      pop1kg_description %>%
        dplyr::filter(SUPER_POP %in% params[["population"]]) %>%
        .[["POP"]]
    ))
}

not_target_population <- dplyr::mutate(not_target_population, QC = "Ethnicity")
```

```{r population-03, fig.height = 4.7 * 1.5, fig.cap = paste("First factorial plane computed with a combined dataset", "involving samples from the 1,000 Genomes Project.", 'Samples in **A** were coloured by "super population" groups,', 'whereas samples in **B** were coloured by "population" groups', "and focused on the PCA area of the ", paste0("**", params[["cohort_name"]], "**"), "cohort.")}
data_pca_plot <- data_pca %>%
  dplyr::left_join(
    y = check_ethnicity %>%
      dplyr::select(IID, ClosestPop) %>%
      dplyr::left_join(
        y = recruit_centre %>%
          dplyr::rename(IID = Sample.Name) %>%
          dplyr::mutate(IID = as.character(IID)) %>%
          dplyr::select(IID, Cohort),
        by = "IID"
      ),
    by = "IID"
  ) %>%
  dplyr::mutate(
    Cohort = ifelse(is.na(Cohort), "1,000 Genomes", Cohort),
    ClosestPop = ifelse(is.na(ClosestPop), as.character(Population), as.character(ClosestPop))
  ) %>%
  dplyr::mutate(
    Population = ifelse(Population == "raw_data", Cohort, as.character(Population))
  ) %>% 
  dplyr::left_join(pop1kg_description, by = c("Population" = "POP")) %>% 
  dplyr::mutate(
    SUPER_POP = ifelse(is.na(SUPER_POP), Population, SUPER_POP),
    SUPER_POP = relevel(x = factor(SUPER_POP), ref = params[["cohort_name"]]),
    SUPER_POP = factor(SUPER_POP, levels = rev(levels(SUPER_POP))),
    Population = relevel(x = factor(Population), ref = params[["cohort_name"]]),
    Population = factor(Population, levels = rev(levels(Population)))
  )

get_lims <- dplyr::filter(data_pca_plot, Population %in% params[["cohort_name"]])

p_superpop <- ggplot2::ggplot(
  data = data_pca_plot, 
  mapping = ggplot2::aes(x = x, y = y, colour = SUPER_POP, shape = SUPER_POP, fill = SUPER_POP)
) +
  ggplot2::geom_hline(yintercept = 0, linetype = 2) +
  ggplot2::geom_vline(xintercept = 0, linetype = 2) +
  ggplot2::geom_point(na.rm = TRUE) +
  ggforce::geom_mark_ellipse(
    alpha = 0.1, 
    con.cap = 0, 
    con.border = "all", 
    expand = ggplot2::unit(2, "mm")
  ) +
  ggplot2::scale_colour_viridis_d(direction = -1) +
  ggplot2::scale_fill_viridis_d(direction = -1) +
  ggplot2::scale_shape_manual(values = rep(c(1:5), 3)) +
  ggplot2::labs(
    x = "First Principal Component", 
    y = "Second Principal Component", 
    colour = NULL, # "Super Population", 
    shape = NULL, # "Super Population", 
    fill = NULL # "Super Population"
  ) +
  ggforce::facet_zoom(
    xlim = range(get_lims[["x"]]), 
    ylim = range(get_lims[["y"]]),
    zoom.size = 0.5,
    horizontal = FALSE
  ) +
  ggplot2::theme(
    legend.position = "bottom", 
    # legend.text = ggplot2::element_text(size = ggplot2::rel(0.5)),
    legend.key.size = ggplot2::unit(0.5, units = "lines")
  )

data_pca_plot_closest <- data_pca_plot %>% 
  dplyr::filter(
    Population %in% c(
      dplyr::filter(data_pca_plot, Population == params[["cohort_name"]])[["ClosestPop"]], 
      params[["cohort_name"]]
    )
  ) %>% 
  dplyr::mutate_if(.predicate = is.factor, .funs = as.character) %>% 
  dplyr::mutate(
    Population = relevel(x = factor(Population), ref = params[["cohort_name"]]),
    Population = factor(Population, levels = rev(levels(Population)))
  )

p_pop <- ggplot2::ggplot(
  data = data_pca_plot_closest,
  mapping = ggplot2::aes(x = x, y = y, colour = Population, shape = Population)
) +
  ggplot2::geom_hline(yintercept = 0, linetype = 2) +
  ggplot2::geom_vline(xintercept = 0, linetype = 2) +
  ggplot2::geom_point(na.rm = TRUE) +
  ggforce::geom_mark_ellipse(
    mapping = ggplot2::aes(fill = Population, colour = Population),
    alpha = 0.1,
    con.cap = 0,
    con.border = "all",
    expand = ggplot2::unit(2, "mm")
  ) +
  ggplot2::scale_colour_viridis_d(direction = -1) +
  ggplot2::scale_fill_viridis_d(direction = -1) +
  ggplot2::scale_shape_manual(values = rep(c(1:5), 3)) +
  ggplot2::labs(
    x = "First Principal Component", 
    y = "Second Principal Component",
    colour = NULL, # "Population", 
    shape = NULL, # "Population", 
    fill = NULL # "Population"
  ) +
  ggplot2::coord_cartesian(
    xlim = range(data_pca_plot_closest[["x"]]), 
    ylim = range(data_pca_plot_closest[["y"]]), 
    expand = 0.05
  ) +
  ggplot2::theme(
    legend.position = "bottom",
    # legend.text = ggplot2::element_text(size = ggplot2::rel(0.5)),
    legend.key.size = ggplot2::unit(0.5, units = "lines")
  )

p_inertia <- inertia_pca %>%
  dplyr::mutate(x = as.numeric(x)) %>%
  dplyr::filter(x %in% 1:params[["pca_components"]]) %>%
  ggplot2::ggplot(mapping = ggplot2::aes(x = x, y = y)) +
  ggplot2::geom_bar(stat = "identity", fill = scales::viridis_pal(begin = 0.5, end = 0.5)(1)) +
  ggplot2::labs(y = "Inertia", x = "Principal Components", title = "PCA Scree Plot") +
  ggplot2::scale_x_continuous(breaks = seq_len(params[["pca_components"]])) +
  ggplot2::scale_y_continuous(
    labels = scales::percent_format(accuracy = 0.01), 
    expand = ggplot2::expand_scale(mult = c(0, 0.05))
  )

ggpubr::ggarrange(
  p_superpop, 
  ggpubr::ggarrange(
    p_pop, 
    p_inertia, 
    nrow = 2, 
    labels = c("B", "C"), 
    heights = c(2, 1)
  ), 
  ncol = 2, 
  labels = c("A", "")
)
```

```{r population-03-bis, fig.height = 4.7 * 1.5, fig.cap = paste("First factorial plane computed with a combined dataset", "involving samples from the 1,000 Genomes Project.", 'Samples were coloured by **A** "super population" and **B** "population" groups.')}
ggpubr::ggarrange(
  p_superpop  +
    ggplot2::facet_grid(cols = ggplot2::vars(Cohort)) +
    ggplot2::theme(legend.position = "top"),
  p_pop +
    ggplot2::facet_grid(cols = ggplot2::vars(Cohort)) +
    ggplot2::theme(legend.position = "top"), 
  nrow = 2, 
  labels = LETTERS
)
```


```{r population-04, eval = !is.null(params[["population"]]) & nrow(not_target_population) > 0}
samples_toexclude <- dplyr::bind_rows(
  samples_toexclude, 
  dplyr::select(not_target_population, FID, IID, QC)
)

pop_name <- pop1kg_description %>%
  dplyr::filter(SUPER_POP %in% params[["population"]]) %>%
  dplyr::distinct(SUPER_POP_NAME) %>%
  .[["SUPER_POP_NAME"]]

if (nrow(not_target_population) > 15) {
  cat(
    "In total, ", scales::comma(nrow(not_target_population)), " samples was(were) determined as ", 
    "non-", pop_name, "(s) based on PCA.\n\n", 
    sep = ""
  )
} else {
  not_target_population %>%
    dplyr::select(FID, IID, ClosestPop) %>%
    dplyr::rename("Closest Population" = ClosestPop) %>%
    pretty_kable(
      align = "c", 
      table.attr = "style='width:65%;'",
      caption = paste0(
        "Samples determined as non-", pop_name, "(s) based on PCA ", 
        "(N = ", scales::comma(nrow(not_target_population)), ")."
      )
    )
}
```

```{r population-05, include = FALSE}
## Only pop studied
samples_toexclude %>%
  dplyr::filter(grepl("Sample_Call_Rate|Heterozygosity_Check$", QC)) %>%
  dplyr::distinct(FID, IID) %>%
  utils::write.table(
    file = paste0(temp_directory, "/samples_toexclude_pca_within.txt"),
    sep = "\t", 
    row.names = FALSE, 
    col.names = FALSE, 
    quote = FALSE
  )

system(command = paste(
  params[["bin_path"]][["plink"]], 
  "--bfile", paste0(temp_directory, "/clean_data_tmp"), 
  "--remove", paste0(temp_directory, "/samples_toexclude_pca_within.txt"), 
  "--make-bed", 
  "--out", paste0(temp_directory, "/clean_data_tmp_within")
), ignore.stdout = TRUE, ignore.stderr = TRUE)

system(command = paste(
  params[["bin_path"]][["gcta"]], 
  "--bfile", paste0(temp_directory, "/clean_data_tmp_within"), 
  "--make-grm-bin", 
  "--thread-num", params[["n_cores"]],
  "--out", paste0(temp_directory, "/clean_data_tmp_within")
), ignore.stdout = TRUE, ignore.stderr = TRUE)

system(command = paste(
  params[["bin_path"]][["gcta"]], 
  "--grm", paste0(temp_directory, "/clean_data_tmp_within"), 
  "--pca", params[["pca_components"]],
  "--thread-num", params[["n_cores"]],
  "--out", paste0(temp_directory, "/clean_data_tmp_within")
), ignore.stdout = TRUE, ignore.stderr = TRUE)

eigenvec_pca <- data.table::fread(paste0(temp_directory, "/clean_data_tmp_within.eigenvec")) %>%
  `colnames<-`(c("FID", "IID", paste0("PC", seq_len(params[["pca_components"]])))) %>%
  dplyr::mutate_at(.vars = c("FID", "IID"), .funs = as.character) %>% 
  dplyr::mutate(
    extremes = abs(PC1) > params[["pca_threshold"]] * stats::sd(PC1) | 
      abs(PC2) > params[["pca_threshold"]] * stats::sd(PC2)
  )

samples_toexclude <- dplyr::bind_rows(
  samples_toexclude,
  eigenvec_pca %>%
    dplyr::filter(extremes) %>%
    dplyr::select(FID, IID) %>%
    dplyr::mutate(QC = "Cohortwithin")
)

recruit_centre_pca <- eigenvec_pca %>%
  dplyr::mutate(POP = "raw_data") %>%
  dplyr::rename(Extremes = extremes) %>%
  dplyr::left_join(
    recruit_centre %>%
      dplyr::rename(IID = Sample.Name) %>%
      dplyr::select(IID, Cohort),
    by = "IID"
  ) %>%
  dplyr::mutate(Cohort = as.factor(Cohort))
```


### Within **`r params[["cohort_name"]]`** structure

To assess the homogeneity of the **`r params[["cohort_name"]]`** cohort, 
a PCA was performed using only samples from the cohort.  
Samples were labeled and flagged as being extreme when the distance from
the **`r params[["cohort_name"]]`** cohort centroid was greater than
`r qdap::replace_number(params[["pca_threshold"]])` times the standard deviation 
of all distances (Figure \@ref(fig:population-06)).

Samples labeled and flagged as extreme: **`r scales::comma(sum(eigenvec_pca[["extremes"]]))`**

```{r population-06, fig.cap = paste("Within-samples outliers detected as samples having at least one coordinate", "in the first factorial plane larger than", qdap::replace_number(params[["pca_threshold"]]), "times the standard deviation of the principal component.")}
p_within <- ggplot2::ggplot(
  data = dplyr::mutate(
    .data = recruit_centre_pca, 
    labs = ifelse(Extremes, IID, NA),
    Extremes = factor(Extremes)
  ),
  mapping = ggplot2::aes(x = PC1, y = PC2)
) +
  ggplot2::geom_hline(yintercept = 0, colour = "black", linetype = 2) +
  ggplot2::geom_vline(xintercept = 0, colour = "black", linetype = 2) +
  ggplot2::geom_point(
    mapping = ggplot2::aes(colour = Extremes, shape = Extremes), 
    na.rm = TRUE
  ) +
  ggrepel::geom_label_repel(
    mapping = ggplot2::aes(label = labs), 
    min.segment.length = ggplot2::unit(0, 'lines'),
    force = 10, 
    alpha = 0.8, 
    size = 2.5,
    na.rm = TRUE
  ) +
  ggplot2::scale_shape_manual(values = c(1, 3)) +
  ggplot2::scale_colour_manual(values = c(scales::viridis_pal(begin = 0.5, end = 0.5)(1), "firebrick2")) +
  ggplot2::labs(
    x = "First Principal Component", 
    y = "Second Principal Component", 
    colour = "Cohort"
  ) +
  ggplot2::theme(legend.position = "none")

## Compute components inertia
inertia_within <- data.table::fread(paste0(temp_directory, "/clean_data_tmp_within.eigenval")) %>%
  tibble::rownames_to_column("x") %>%
  dplyr::mutate(y = V1 / sum(V1)) %>%
  dplyr::select(x, y) %>% 
  dplyr::mutate(x = as.numeric(x))

p_inertia_within <- ggplot2::ggplot(
  data = dplyr::filter(.data = inertia_within, x %in% 1:params[["pca_components"]]),
  mapping = ggplot2::aes(x = x, y = y)
) +
  ggplot2::geom_bar(stat = "identity", fill = scales::viridis_pal(begin = 0.5, end = 0.5)(1)) +
  ggplot2::labs(y = "Inertia", x = "Principal Components", title = "PCA Scree Plot") +
  ggplot2::scale_x_continuous(breaks = seq_len(params[["pca_components"]])) +
  ggplot2::scale_y_continuous(
    labels = scales::percent_format(accuracy = 0.01), 
    expand = ggplot2::expand_scale(mult = c(0, 0.05))
  )

ggpubr::ggarrange(p_within, p_inertia_within, ncol = 2, labels = LETTERS, widths = c(1.2, 1))

pca_threshold <- which(abs(diff(inertia_within[["y"]])) < 0.01)
```

<!--According to the eigenvalues, the first factorial plane explains 
`r scales::percent(sum(inertia_within[c(1, 2), "y"]), accuracy = 0.01)` of the variance among samples. 
The principal components after `r qdap::replace_number(pca_threshold[1])` explain on average 
`r scales::percent(mean(inertia_within[pca_threshold[1]:nrow(inertia_within), "y"]), accuracy = 0.01)` 
[`r paste(scales::percent(range(inertia_within[pca_threshold[1]:nrow(inertia_within), "y"]), accuracy = 0.01), collapse = "; ")`] of 
the variance (principal components up to `r  scales::comma(nrow(inertia_within))`).-->

```{r population-07, fig.cap = "PCA plot for components from two to five."}
pca_planes <- utils::combn(seq_len(min(params[["pca_components"]], 5))[-1], 2)

pca_planes_list <- lapply(seq(ncol(pca_planes)), function(iComb) {
  ggplot2::ggplot(data = recruit_centre_pca) +
    ggplot2::geom_hline(yintercept = 0, colour = "black", linetype = 2) +
    ggplot2::geom_vline(xintercept = 0, colour = "black", linetype = 2) +
    ggplot2::geom_point(
      ggplot2::aes(
        x = .data[[paste0("PC", pca_planes[1, iComb])]], 
        y = .data[[paste0("PC", pca_planes[2, iComb])]], 
        colour = Extremes, 
        shape = Extremes
      ), 
      shape = 1, 
      na.rm = TRUE
    ) +
    ggplot2::scale_shape_manual(values = c(1, 3)) +
    ggplot2::scale_colour_manual(values = c(scales::viridis_pal(begin = 0.5, end = 0.5)(1), "firebrick2")) +
    ggplot2::theme(legend.position = "none") +
    ggplot2::labs(
      x = paste0("PC", pca_planes[1, iComb]),
      y = paste0("PC", pca_planes[2, iComb])
    )
})

ggpubr::ggarrange(plotlist = pca_planes_list, nrow = 2, ncol = 3, labels = LETTERS)
```

```{r exclusion, include = FALSE}
samples_toexclude_export <- samples_toexclude %>%
  dplyr::group_by(QC) %>% 
  dplyr::distinct(FID, IID, .keep_all = TRUE) %>% 
  dplyr::ungroup() %>% 
  dplyr::mutate(
    QC = factor(
      x = QC,
      levels = c(
        "Sample_Call_Rate", "Gender_Discrepancy", "Gender_Missing", 
        "Gender_F", "Heterozygosity_Check", 
        paste0("Heterozygosity_Check_maf_", params[["maf_threshold"]]),
        "Relatedness", "Sample_Mendel_Error", "Ethnicity", "Cohortwithin"
      )
    ),
    value = 1
  ) %>%
  tidyr::complete(tidyr::nesting(FID, IID), QC, fill = list(value = 0)) %>% 
  tidyr::spread(QC, value) %>% 
  dplyr::mutate(
    Status = dplyr::if_else((Sample_Call_Rate + Heterozygosity_Check) > 0, "Exclude", "Check")
  ) %>% 
  dplyr::rename(
    `Call Rate` = Sample_Call_Rate,
    `Gender Discrepancy` = Gender_Discrepancy,
    `Gender Missing` = Gender_Missing,
    `Gender F` = Gender_F,
    !!paste0("Heterozygosity Check (MAF ", params[["maf_threshold"]], ")") :=
      paste0("Heterozygosity_Check_maf_", params[["maf_threshold"]]),
    `Heterozygosity Check` = Heterozygosity_Check,
    `PCA Outliers` = Ethnicity,
    `Within Cohort PCA outliers` = Cohortwithin,
    `Relatedness Pi_hat ≥ 0.2` = Relatedness,
    `Mendelian Error` = Sample_Mendel_Error
  )

utils::write.table(
  x = dplyr::filter(samples_toexclude_export, Status == "Exclude"),
  file = paste0(temp_directory, "/samples_excluded.txt"),
  sep = "\t", 
  col.names = TRUE, 
  quote = FALSE, 
  row.names = FALSE
)

system(command = paste(
  params[["bin_path"]][["plink"]], 
  "--bfile", paste0(temp_directory, "/raw_data"), 
  "--remove", paste0(temp_directory, "/samples_excluded.txt"), 
  "--snps-only", 
  "--set-hh-missing", 
  "--make-bed", 
  "--out", paste0(temp_directory, "/raw_data_good_samples")
), ignore.stdout = TRUE, ignore.stderr = TRUE)
```



# Variant-based QC

After excluding sample-based QC failed samples, variant-based QC was performed on 
**`r scales::comma(nrow(data.table::fread(input = paste0(temp_directory, "/raw_data_good_samples.fam"), select = 1)))`** samples.  


## Variant call rate

```{r variant-call-rate-01, include = FALSE}
system(command = paste(
  params[["bin_path"]][["plink"]], 
  "--bfile", paste0(temp_directory, "/raw_data_good_samples"), 
  "--missing",
  "--out ", paste0(temp_directory, "/raw_data_good_samples")
), ignore.stdout = TRUE, ignore.stderr = TRUE)

callrate_snps_raw <- data.table::fread(paste0(temp_directory, "/raw_data_good_samples.lmiss"))
```

Proportion of missing genotypes or call rate per variant was estimated using the `--missing` command in PLINK.  

```{r variant-call-rate-02, fig.cap = paste("Call rate per variant. The red line denotes the call rate", scales::percent(params[["callrate_snps"]], accuracy = 0.01),  "cut-off.", "Variants with a call rate greater than or equal to 99.9% are not shown.")}
ggplot2::ggplot(
  data = callrate_snps_raw %>%
    dplyr::arrange(F_MISS) %>%
    tidyr::replace_na(replace = list(F_MISS = 1)) %>% 
    dplyr::filter(F_MISS>=0.001),
  mapping = ggplot2::aes(x = length(F_MISS):1, y = 1 - F_MISS)
) +
  ggplot2::geom_point(
    colour = scales::viridis_pal(begin = 0.5, end = 0.5)(1), 
    shape = 1, 
    na.rm = TRUE
  ) +
  ggplot2::geom_hline(
    mapping = ggplot2::aes(yintercept = params[["callrate_snps"]]),
    colour = "firebrick2", 
    linetype = 2
  ) +
  ggplot2::scale_x_continuous(labels = scales::comma_format(), trans = "log10") +
  ggplot2::scale_y_continuous(labels = scales::percent_format(accuracy = 0.01)) +
  ggplot2::labs(x = "Variants", y = "Call rate")
```


## Duplicated and/or triplicated variants

```{r duplicated-snps, include = FALSE}
callrate_snps <- callrate_snps_raw %>%
  dplyr::select(CHR, SNP, F_MISS) %>%
  dplyr::left_join(
    y = data.table::fread(paste0(temp_directory, "/raw_data_good_samples.bim")) %>%
      dplyr::select(CHR = V1, SNP = V2, POS = V4),
    by = c("CHR" = "CHR", "SNP" = "SNP")
  ) %>%
  dplyr::arrange(CHR, POS) %>%
  tidyr::unite(CHRPOS, CHR, POS, sep = ":", remove = FALSE)

duplicated_snps <- callrate_snps %>%
  dplyr::filter(CHRPOS %in% CHRPOS[duplicated(CHRPOS)])

if (nrow(duplicated_snps) == 0) {
  uniq_snps <- callrate_snps
} else {
  uniq_snps <- dplyr::bind_rows(
    callrate_snps %>%
      dplyr::filter(!CHRPOS %in% duplicated_snps[["CHRPOS"]]),
    duplicated_snps %>% 
      dplyr::group_by(CHRPOS) %>% 
      dplyr::do(
        dplyr::filter(., F_MISS == min(F_MISS))
      ) %>%
      dplyr::filter(dplyr::n()==1) %>% 
      dplyr::ungroup()
  ) %>% 
    dplyr::mutate_at(.vars = c("CHR", "POS"), as.character)
}
```

Due to some variants having several different names, 
several variants with two or more identical copies were found in the dataset. 
For each variant with multiple copies, 
only the variant with the highest call rate (or arbitrarily if call rate was equal) was included.  

In total, **`r scales::comma(nrow(callrate_snps) - nrow(uniq_snps))`** variants 
that possibly existed in multiple copies in the dataset (same chromosome and position) were removed.  


## Variant Mendel errors

```{r lmendel-error-00}
if (!params[["includes_relatives"]]) cat("Not Applicable.\n\n")

lmendel_file_check <- params[["includes_relatives"]]
```

```{r lmendel-error-01, eval = params[["includes_relatives"]]}
system(command = paste(
  params[["bin_path"]][["plink"]], 
  "--bfile", paste0(temp_directory, "/raw_data_good_samples"), 
  "--mendel",
  "--out ", paste0(temp_directory, "/raw_data_good_samples")
), ignore.stdout = TRUE, ignore.stderr = TRUE)

lmendel_file_check <- file.exists(paste0(temp_directory, "/raw_data_good_samples.lmendel"))

cat(
  "The variant-wise Mendel error was estimated using the `--mendel` command in PLINK.  \n",
  "The per-variant error rate threshold is `", params[["mendelian_snp"]], "`.\n\n", 
  sep = ""
)

if (!lmendel_file_check) {
  cat("Mendel error rates check was not performed due to missing informations.\n\n")
} else {
  mendel_error_snps <- data.table::fread(paste0(temp_directory, "/raw_data_good_samples.lmendel")) %>%
    dplyr::mutate(percentage = N / dplyr::n()) %>%
    dplyr::arrange(percentage)
}
```

```{r lmendel-error-02, eval = params[["includes_relatives"]] & lmendel_file_check, fig.cap = paste("Mendel error rate per variant.", scales::comma(sum(mendel_error_snps[["percentage"]] > params[["mendelian_snp"]])), "variants with more than", scales::percent(x = params[["mendelian_snp"]], accuracy = 0.01), "Mendel errors rate were excluded.")}
ggplot2::ggplot(
  data = mendel_error_snps,
  mapping = ggplot2::aes(x = seq_len(percent), y = percent)
) +
  ggplot2::geom_point(
    colour = scales::viridis_pal(begin = 0.5, end = 0.5)(1), 
    shape = 1, 
    na.rm = TRUE
  ) +
  ggplot2::geom_hline(yintercept = params[["mendelian_snp"]], colour = "firebrick2", linetype = 2) +
  ggplot2::scale_x_continuous(labels = scales::comma_format(), trans = "log10") +
  ggplot2::scale_y_continuous(labels = scales::percent_format(accuracy = 0.01)) +
  ggplot2::labs(x = "Variants", y = "Mendel Error Rate")
```

```{r uniq-snps-list, include = FALSE}
if (lmendel_file_check) {
  cat(
    setdiff(
      uniq_snps[["SNP"]], 
      dplyr::filter(mendel_error_snps, percentage > params[["mendelian_snp"]])[["SNP"]]
    ),
    sep = "\n", 
    file = paste0(temp_directory, "/uniq_snps_list.txt")
  )
} else {
  cat(
    uniq_snps[["SNP"]], 
    sep = "\n", 
    file = paste0(temp_directory, "/uniq_snps_list.txt")
  )
}
```


## Minor allele frequency distribution

The MAF were computed using the `--freq` command in PLINK.

```{r maf-01, include = FALSE}
system(command = paste(
  params[["bin_path"]][["plink"]], 
  "--bfile", paste0(temp_directory, "/raw_data_good_samples"), 
  "--extract", paste0(temp_directory, "/uniq_snps_list.txt"),
  "--freq",
  "--out", paste0(temp_directory, "/raw_data_good_samples")
), ignore.stdout = TRUE, ignore.stderr = TRUE)
```

```{r maf-02, fig.cap = "Distribution of variants number per class of minor allele frequency (MAF)"}
data.table::fread(paste0(temp_directory, "/raw_data_good_samples.frq")) %>%
  dplyr::mutate(
    maf_dist = cut(MAF, breaks = c(0, 0.01, 0.05, 0.1, 0.2, 0.3, 0.4, 0.5), include.lowest = TRUE)
  ) %>% 
  tidyr::drop_na(MAF) %>% 
  ggplot2::ggplot(
    mapping = ggplot2::aes(x = maf_dist)
  ) +
    ggplot2::geom_bar(fill = scales::viridis_pal(begin = 0.5, end = 0.5)(1)) +
    ggplot2::labs(x = "MAF", y = "Number of variants") +
    ggplot2::scale_y_continuous(
      labels = scales::comma_format(),
      expand = ggplot2::expand_scale(mult = c(0, 0.1))
    )
```


## Hardy-Weinberg equilibrium

```{r hwe-01, include = FALSE}
system(command = paste(
  params[["bin_path"]][["plink"]], 
  "--bfile", paste0(temp_directory, "/raw_data_good_samples"),
  "--hardy", 
  "--out ", paste0(temp_directory, "/raw_data_good_samples")
), ignore.stdout = TRUE, ignore.stderr = TRUE)

hwe_complete <- data.table::fread(paste0(temp_directory, "/raw_data_good_samples.hwe")) %>%
  tidyr::drop_na(P)

hwe_significant <- hwe_complete %>%
  dplyr::filter(P < params[["hwe_pvalue"]] * 10) %>%
  dplyr::arrange(P)
```

Hardy-Weinberg equilibrium (HWE) p-value was computed using the `--hardy` command in PLINK.

In total, **`r scales::comma(sum(hwe_complete[, "P"] < params[["hwe_pvalue"]]))`** variants 
had a HWE p-value below `r scales::scientific(params[["hwe_pvalue"]])` and were excluded.  

```{r hwe-02, fig.cap = paste0("The distribution of Hardy-Weinberg Equilibrium (HWE) p-values ",  "for variants with HWE p-value < ", scales::scientific(params[["hwe_pvalue"]] * 10), ". ", "The horizontal red line denotes the ", scales::scientific(params[["hwe_pvalue"]]), " cut-off.")}
ggplot2::ggplot(data = hwe_significant) +
  ggplot2::geom_point(
    mapping = ggplot2::aes(x = 1:length(P), y = P),
    colour = scales::viridis_pal(begin = 0.5, end = 0.5)(1), 
    shape = 1, 
    na.rm = TRUE
  ) +
  ggplot2::geom_hline(yintercept = params[["hwe_pvalue"]], colour = "firebrick2", linetype = 2) +
  ggplot2::scale_x_continuous(labels = scales::comma_format(), trans = "log10") +
  ggplot2::scale_y_continuous(trans = pval_trans()) +
  ggplot2::labs(x = "Variants", y = "Hardy-Weinberg Equilibrium P-value")
```

```{r file-management, include = FALSE}
imputation_filename <- paste0(temp_directory, "/", params[["cohort_name"]], "_forimputation")

system(command = paste(
  params[["bin_path"]][["plink"]], 
  "--bfile", paste0(temp_directory, "/raw_data"), 
  "--set-hh-missing", 
  "--remove", paste0(temp_directory, "/samples_excluded.txt"),
  "--geno", 1 - params[["callrate_snps"]],
  "--extract", paste0(temp_directory, "/uniq_snps_list.txt"), 
  "--hwe", params[["hwe_pvalue"]],
  "--make-bed", 
  "--out", imputation_filename
))

if (file.exists(paste0(imputation_filename, ".hh"))) {
  system(command = paste(
    params[["bin_path"]][["plink"]], 
    "--bfile", imputation_filename, 
    "--exclude", paste0(imputation_filename, ".hh"), 
    "--make-bed", 
    "--out", imputation_filename
  ))
} else {
  system(command = paste(
    params[["bin_path"]][["plink"]], 
    "--bfile", imputation_filename, 
    "--make-bed", 
    "--out", imputation_filename
  ))
}

system(command = paste(
  params[["bin_path"]][["plink"]], 
  "--bfile", imputation_filename, 
  "--freq", 
  "--out", imputation_filename
))

system(command = paste(
  "(cd", temp_directory, "&&",
    params[["check_bim_script"]],
    "-b", paste0(imputation_filename, ".bim"),
    "-f", paste0(imputation_filename, ".frq"),
    "-r", params[["ref1kg_legend"]], 
    "-g", 
    "--pop ALL",
  ")"
))

system(
  command = paste(
    paste(
      "(cd", temp_directory, ";",
        gsub("plink", params[["bin_path"]][["plink"]], readLines(paste0(temp_directory, "/Run-plink.sh"))),
      ")"
    ),
    collapse = ";"
  )
)

system(command = paste(
  params[["bin_path"]][["plink"]], 
  "--bfile", paste0(imputation_filename, "-updated"), 
  "--reference-allele", 
      paste0(temp_directory, "/Force-Allele1-", params[["cohort_name"]], "_forimputation-1000G.txt"), 
  "--recode vcf",
  "--out", imputation_filename
))

system(command = paste(
  params[["bin_path"]][["bgzip"]], 
  "-c", paste0(imputation_filename, ".vcf"), 
  ">", paste0(imputation_filename, ".vcf.gz")
))

system(command = paste(
  params[["bin_path"]][["bcftools"]], "index", paste0(imputation_filename, ".vcf.gz")
))

system(command = paste(
  params[["bin_path"]][["plink"]], 
  "--bfile", imputation_filename, 
  "--set-hh-missing", 
  "--remove ", paste0(temp_directory, "/samples_excluded.txt"), 
  "--geno", 1 - params[["callrate_snps"]], 
  "--extract", paste0(temp_directory, "/uniq_snps_list.txt"),
  "--hwe", params[["hwe_pvalue"]],
  "--make-bed", 
  "--out", paste0(params[["output_directory"]], "/plink_qc/", params[["cohort_name"]], "_QC")
))

utils::write.table(
  x = structure(
    list(
      V1 = 1:26,
      V2 = structure(c(
        1L, 12L, 16L, 17L, 18L, 19L, 20L, 21L, 22L, 2L, 3L, 4L, 5L, 6L, 
        7L, 8L, 9L, 10L, 11L, 13L, 14L, 15L, 24L, 25L, 24L, 23L
      ),
      .Label = c(
        "1", "10", "11", "12", "13", "14", "15", "16", "17", "18", "19", "2",
        "20", "21", "22", "3", "4", "5", "6", "7", "8", "9", "MT", "X", "Y"
      ),
      class = "factor"
      )
    ),
    .Names = c("V1", "V2"), class = "data.frame", row.names = c(NA, -26L)
  ),
  file = paste0(temp_directory, "/plink2ensembl.txt"),
  row.names = FALSE, 
  col.names = FALSE, 
  quote = FALSE, 
  sep = "\t"
)

system(command = paste(
  params[["bin_path"]][["bcftools"]], "annotate", 
  "--rename-chrs", paste0(temp_directory, "/plink2ensembl.txt"),
  paste0(imputation_filename, ".vcf.gz"),
  "-Oz", 
  ">", paste0(imputation_filename, "-tmp.vcf.gz")
))

system(command = paste(
  params[["bin_path"]][["bcftools"]], "+fixref", 
  paste0(imputation_filename, "-tmp.vcf.gz"), 
  "-Ob -o", paste0(params[["output_directory"]], "/vcf_qc/", params[["cohort_name"]], "_QC.vcf.gz"),
  "--", 
  "-d", 
  "-f", params[["ref1kg_fasta"]], 
  "-m flip"
))

system(command = paste(
  params[["bin_path"]][["bcftools"]], "norm", 
  "--check-ref e", 
  "-f ", params[["ref1kg_fasta"]], 
  paste0(params[["output_directory"]], "/vcf_qc/", params[["cohort_name"]], "_QC.vcf.gz"), 
  "-o /dev/null"
))

system(command = paste(
  params[["bin_path"]][["bcftools"]], "index", 
  paste0(params[["output_directory"]], "/vcf_qc/", params[["cohort_name"]], "_QC.vcf.gz")
))
```



# Final QC overview (with all filters)

```{r qc-summary, include = FALSE}
initial_snps_number <- nrow(data.table::fread(
  input = paste0(params[["input_directory"]], ".bim"), 
  select = 1
))

final_snps_number <- nrow(data.table::fread(
  input = paste0(params[["output_directory"]], "/plink_qc/", params[["cohort_name"]], "_QC.bim"), 
  select = 1
))

initial_samples_number <- nrow(data.table::fread(
  input = paste0(params[["input_directory"]], ".fam"),
  select = 1
))

final_samples_number <- nrow(data.table::fread(
  input = paste0(params[["output_directory"]], "/plink_qc/", params[["cohort_name"]], "_QC.fam"), 
  select = 1
))
```

* Samples previously characterised in Section [Sample-based QC](#sample-qc) (Table \@ref(tab:qc-summary-table)):
    + **`r scales::comma(dplyr::n_distinct(dplyr::filter(samples_toexclude_export, Status=="Exclude")[["IID"]]))`**
        samples excluded due to failures in the call rate and/or the heterozygosity checks.
    + **`r scales::comma(dplyr::n_distinct(dplyr::filter(samples_toexclude_export, Status=="Check")[["IID"]]))`**
        samples flagged. 
        Those samples need to be checked in order to determine whether or not they have to be excluded.
* Variant call rate at `r scales::percent(params[["callrate_snps"]], accuracy = 0.01)`.
* Duplicated variants based on chromosome and position
* Hardy-Weinberg Equilibrium p-value at `r scales::scientific(params[["hwe_pvalue"]])`
`r if (params[["includes_relatives"]] & lmendel_file_check) cat("* Per-variant Mendelian error rate (when applicable) at", params[["mendelian_snp"]])`

The initial dataset included **`r scales::comma(initial_samples_number)`** samples 
and **`r scales::comma(initial_snps_number)`** variants.  
The final dataset includes **`r scales::comma(final_samples_number)`** samples 
and **`r scales::comma(final_snps_number)`** variants.

```{r qc-summary-table}
recruit_centre %>%
  dplyr::select(Sample.Name, Cohort, Gender) %>%
  dplyr::right_join(
    y = dplyr::select(samples_toexclude_export, IID, Status), 
    by = c("Sample.Name" = "IID")
  ) %>% 
  dplyr::group_by(Cohort, Status, .drop = FALSE) %>%
  dplyr::count() %>%
  tidyr::spread(key = Status, value = n) %>% 
  dplyr::left_join(
    y = dplyr::select(recruit_centre_table, Cohort, Total), 
    by = "Cohort"
  ) %>%
  dplyr::rowwise() %>% 
  dplyr::mutate(
    ExcludeorCheck_pct = scales::percent((Exclude + Check) / Total, accuracy = 0.01),
    ExcludeorCheck = paste0((Exclude + Check), " / ", Total, " (", ExcludeorCheck_pct, ")") %>% gsub("NaN", "0", .),
    Check_pct = scales::percent(Check / Total, accuracy = 0.01), 
    Exclude_pct = scales::percent(Exclude / Total, accuracy = 0.01),
    Check = paste0(Check, " / ", Total, " (", Check_pct, ")") %>% gsub("NaN", "0", .),
    Exclude = paste0(Exclude, " / ", Total, " (", Exclude_pct, ")") %>% gsub("NaN", "0", .)
  ) %>%
  dplyr::select(
    Cohort, 
    Excluded = Exclude, 
    Flagged = Check, 
    `Excluded or Flagged` = ExcludeorCheck
  ) %>%
  pretty_kable(
    align = "c", 
    table.attr = "style='width:65%;'",
    caption = "Excluded samples and flagged samples per cohort."
  )
```

```{r output-exclusion, include = FALSE}
output_xlsx <- list()

output_xlsx[["exclude_flag"]] <- samples_toexclude_export %>%
  dplyr::left_join(
    y = recruit_centre %>%
      dplyr::select(Sample.Name, Cohort, Gender) %>%
      dplyr::mutate(Sample.Name = as.character(Sample.Name)),
    by = c("IID" = "Sample.Name")
  ) %>%
  dplyr::rename(
    "Ethnicity" = "PCA Outliers",
    "Cohort PCA outliers" = "Within Cohort PCA outliers"
  ) %>%
  dplyr::select(FID, IID, Gender, Cohort, Status, dplyr::everything()) %>% 
  dplyr::arrange(dplyr::desc(!!as.symbol(colnames(.)[-c(1:4)])))

if (exists("related_samples")) {
  output_xlsx[["relatedness"]] <- related_samples %>%
    dplyr::left_join(
      y = dplyr::select(recruit_centre, Sample.Name, Cohort),
      by = c("IID1" = "Sample.Name")
    ) %>%
    dplyr::left_join(
      y = dplyr::select(recruit_centre, Sample.Name, Cohort),
      by = c("IID2" = "Sample.Name"),
      suffix = c("1", "2")
    )
}

if (exists("mendel_errors_samples")) {
  output_xlsx[["Sample_Mendel_Error_Rate"]] <- mendel_errors_samples %>% 
    dplyr::mutate(
      `Sample Mendel error rate` = scales::percent(percent, accuracy = 0.01)
    ) %>% 
    dplyr::select(-percent)
}

if (nrow(not_target_population) != 0) {
  output_xlsx[["ethnicity"]] <- dplyr::select(not_target_population, FID, IID, ClosestPop)
}

writexl::write_xlsx(output_xlsx, path = paste0(params[["output_directory"]], "/QC_exclusion.xlsx"))
```



# Technical information


## Softwares

```{r softwares-version, include = FALSE}
plink_ver <- strsplit(
  x = system(command = paste(params[["bin_path"]][["plink"]], "--version"), intern = TRUE)[1],
  split = " "
)[[1]][2]
gcta_ver <- paste(strsplit(
  x = suppressWarnings(system(command = params[["bin_path"]][["gcta"]], intern = TRUE)[3]),
  split = " "
)[[1]][3:4], collapse = "-")
bgzip_ver <- strsplit(
  x = system(command = paste(params[["bin_path"]][["bgzip"]], "--version"), intern = TRUE)[1],
  split = " "
)[[1]][3]
bcftools_ver <- strsplit(
  x = system(command = paste(params[["bin_path"]][["bcftools"]], "--version"), intern = TRUE)[1],
  split = " "
)[[1]][2]
check_bim_script_ver <- readLines(params[["check_bim_script"]], warn = FALSE) %>% 
  grep("^#", ., value = TRUE) %>% 
  grep("-v[0-9]", ., value = TRUE) %>% 
  gsub("# *-", "", .) %>% 
  sort(decreasing = TRUE) %>% 
  .[1]
```

* plink `r plink_ver` (Chang et al., 2015; doi:10.1186/s13742-015-0047-8)
* gcta `r gcta_ver` (Yang et al., 2011; doi:10.1016/j.ajhg.2010.11.011)
* bgzip `r bgzip_ver` ([http://www.htslib.org/](http://www.htslib.org/))
* bcftools `r bcftools_ver` ([http://www.htslib.org/](http://www.htslib.org/))
* `r basename(params[["check_bim_script"]])` `r check_bim_script_ver` (William Rayner, [https://www.well.ox.ac.uk/~wrayner/tools/](https://www.well.ox.ac.uk/~wrayner/tools/))


## R session information

```{r session-info, results = "markup"}
options("width" = 110)
sessioninfo::session_info()
```
