<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>qc_plink — qc_plink • CARoT</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/simplex/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="qc_plink — qc_plink" />

<meta property="og:description" content="qc_plink" />
<meta name="twitter:card" content="summary" />

<meta name="robots" content="noindex">

<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">CARoT</a>
        <span class="version label label-danger" data-toggle="tooltip" data-placement="bottom" title="In-development version">0.2.1.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>qc_plink</h1>
    
    <div class="hidden name"><code>qc_plink.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>qc_plink</p>
    
    </div>

    <pre class="usage"><span class='fu'>qc_plink</span>(<span class='kw'>input_directory</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>output_directory</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>cohort_name</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>output_file</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/paste'>paste0</a></span>(<span class='no'>cohort_name</span>, <span class='st'>"_QC.html"</span>),
  <span class='kw'>array</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>callrate_samples</span> <span class='kw'>=</span> <span class='fl'>0.95</span>, <span class='kw'>callrate_snps</span> <span class='kw'>=</span> <span class='fl'>0.95</span>,
  <span class='kw'>heterozygosity_treshold</span> <span class='kw'>=</span> <span class='fl'>4</span>, <span class='kw'>maf_threshold</span> <span class='kw'>=</span> <span class='fl'>0.01</span>,
  <span class='kw'>hwe_pvalue</span> <span class='kw'>=</span> <span class='fl'>1e-04</span>, <span class='kw'>includes_relatives</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>mendelian_samples</span> <span class='kw'>=</span> <span class='fl'>0.05</span>, <span class='kw'>mendelian_snp</span> <span class='kw'>=</span> <span class='fl'>0.1</span>, <span class='kw'>IBD_threshold</span> <span class='kw'>=</span> <span class='fl'>0.2</span>,
  <span class='kw'>population</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>pca_components</span> <span class='kw'>=</span> <span class='fl'>10</span>, <span class='kw'>pca_threshold</span> <span class='kw'>=</span> <span class='fl'>3</span>,
  <span class='kw'>check_bim_script</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/system.file'>system.file</a></span>(<span class='st'>"perl"</span>, <span class='st'>"HRC-1000G-check-bim.pl"</span>,
  <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>"CARoT"</span>), <span class='kw'>ref1kg_panel</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>ref1kg_population</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>ref1kg_genotypes</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>ref1kg_legend</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>ref1kg_fasta</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>bin_path</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/list'>list</a></span>(<span class='kw'>bcftools</span> <span class='kw'>=</span> <span class='st'>"/usr/bin/bcftools"</span>, <span class='kw'>bgzip</span> <span class='kw'>=</span>
  <span class='st'>"/usr/bin/bgzip"</span>, <span class='kw'>plink</span> <span class='kw'>=</span> <span class='st'>"/usr/bin/plink1.9"</span>, <span class='kw'>gcta</span> <span class='kw'>=</span> <span class='st'>"/usr/bin/gcta64"</span>),
  <span class='kw'>title</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/paste'>paste</a></span>(<span class='no'>array</span>, <span class='st'>"Array Quality-Control"</span>), <span class='kw'>author_name</span> <span class='kw'>=</span> <span class='st'>"CARoT"</span>,
  <span class='kw'>author_affiliation</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>author_email</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>cache</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>show_code</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>n_cores</span> <span class='kw'>=</span> <span class='fl'>22</span>, <span class='kw'>dpi</span> <span class='kw'>=</span> <span class='fl'>120</span>, <span class='kw'>gg_fontsize</span> <span class='kw'>=</span> <span class='fl'>12</span>,
  <span class='kw'>encoding</span> <span class='kw'>=</span> <span class='st'>"UTF-8"</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>input_directory</th>
      <td><p>A <code>character</code>. The path to the plink files.
The path should contains the file name without the extension, i.e., without <code>*.bed</code>, <code>*.bim</code> or <code>*.fam</code>.</p></td>
    </tr>
    <tr>
      <th>output_directory</th>
      <td><p>A <code>character</code>. The path to the output directory.</p></td>
    </tr>
    <tr>
      <th>cohort_name</th>
      <td><p>A <code>character</code>. The name of the studied cohort / population.</p></td>
    </tr>
    <tr>
      <th>output_file</th>
      <td><p>A <code>character</code>. The name of the html file produced.</p></td>
    </tr>
    <tr>
      <th>array</th>
      <td><p>A <code>character</code>. The array name, e.g., "Illumina Omni2.5".</p></td>
    </tr>
    <tr>
      <th>callrate_samples</th>
      <td><p>A <code>numeric</code>. The call rate threshold for samples, under which samples are excluded.
Default is <code>0.95</code>.</p></td>
    </tr>
    <tr>
      <th>callrate_snps</th>
      <td><p>A <code>numeric</code>. The call rate threshold for probes, under which probes are excluded.
Default is <code>0.95</code>.</p></td>
    </tr>
    <tr>
      <th>heterozygosity_treshold</th>
      <td><p>A <code>numeric</code>. The heterozygosity threshold for samples
(number of standard deviation from the mean), under/above which samples are excluded.
Default is <code>4</code>.</p></td>
    </tr>
    <tr>
      <th>maf_threshold</th>
      <td><p>A <code>numeric</code>. The minor allele frequency under which variants are considered "rare".
Default is <code>0.01</code>.</p></td>
    </tr>
    <tr>
      <th>hwe_pvalue</th>
      <td><p>A <code>numeric</code>. The p-value threshold for Hardy-Weinberg equilibrium test.
Default is <code>0.0001</code>.</p></td>
    </tr>
    <tr>
      <th>includes_relatives</th>
      <td><p>A <code>logical</code>. Does the data contain related samples?
Default is <code>FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>mendelian_samples</th>
      <td><p>A <code>numeric</code>. The Mendel error rate threshold above which samples are excluded.
Default is <code>0.05</code>.</p></td>
    </tr>
    <tr>
      <th>mendelian_snp</th>
      <td><p>A <code>numeric</code>. The Mendel error rate threshold above which variants are excluded.
Default is <code>0.1</code>.</p></td>
    </tr>
    <tr>
      <th>IBD_threshold</th>
      <td><p>A <code>numeric</code>. The threshold for IBD (identical by descent) above which
samples are characterised as relatives.
Default is <code>0.2</code>.</p></td>
    </tr>
    <tr>
      <th>population</th>
      <td><p>A <code>character</code>. The ethnicity of the studied population if known, e.g., <code>"EUR"</code>.
Default is <code>NULL</code>.</p></td>
    </tr>
    <tr>
      <th>pca_components</th>
      <td><p>A <code>numeric</code>. The number of principal components to be computed.
Default is <code>10</code>.</p></td>
    </tr>
    <tr>
      <th>pca_threshold</th>
      <td><p>A <code>numeric</code>. The threshold to define outliers on the principal component analysis,
the as number of standard deviation from the cohort centroid.
Default is <code>3</code>.</p></td>
    </tr>
    <tr>
      <th>check_bim_script</th>
      <td><p>A <code>character</code>. The PERL script to use to check PLINK files to allow later imputation.
Default is <code><a href='https://www.rdocumentation.org/packages/base/topics/system.file'>system.file("perl", "HRC-1000G-check-bim.pl", package = "CARoT")</a></code>.</p></td>
    </tr>
    <tr>
      <th>ref1kg_panel</th>
      <td><p>A <code>character</code>. The <code>*.panel</code> file from 1,000 Genome project.
Default is <code>NULL</code>.</p></td>
    </tr>
    <tr>
      <th>ref1kg_population</th>
      <td><p>A <code>character</code>. The <code>*.tsv</code> file from 1,000 Genome project describing samples and ethnicity.
Default is <code>NULL</code>.</p></td>
    </tr>
    <tr>
      <th>ref1kg_genotypes</th>
      <td><p>A <code>character</code>. The PLINK files from 1,000 Genome project.
Default is <code>NULL</code>.</p></td>
    </tr>
    <tr>
      <th>ref1kg_legend</th>
      <td><p>A <code>character</code>. The <code>*.legend</code> file from 1,000 Genome project.
Default is <code>NULL</code>.</p></td>
    </tr>
    <tr>
      <th>ref1kg_fasta</th>
      <td><p>A <code>character</code>. The <code>*.fasta</code> file from 1,000 Genome project.
Default is <code>NULL</code>.</p></td>
    </tr>
    <tr>
      <th>bin_path</th>
      <td><p>A <code><a href='https://www.rdocumentation.org/packages/base/topics/list'>list(character)</a></code>. A list giving the binary path of <code>bcftools</code>, <code>bgzip</code>, <code>gcta</code> and <code>plink1.9</code>.</p></td>
    </tr>
    <tr>
      <th>title</th>
      <td><p>A <code>character</code>. The report's title. Default is <code><a href='https://www.rdocumentation.org/packages/base/topics/paste'>paste(array, "Array Quality-Control")</a></code>.</p></td>
    </tr>
    <tr>
      <th>author_name</th>
      <td><p>A <code>character</code>. The author's name to be printed in the report.
Default is <code>CARoT</code>.</p></td>
    </tr>
    <tr>
      <th>author_affiliation</th>
      <td><p>A <code>character</code>. The affiliation to be printed in the report.
Default is <code>NULL</code>.</p></td>
    </tr>
    <tr>
      <th>author_email</th>
      <td><p>A <code>character</code>. The email to be printed in the report.
Default is <code>NULL</code>.</p></td>
    </tr>
    <tr>
      <th>cache</th>
      <td><p>A <code>logical</code>. Should the R code be cached?
Default is <code>FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>show_code</th>
      <td><p>A <code>logical</code>. Should the R code be printed in the report?
Default is <code>FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>n_cores</th>
      <td><p>A <code>numeric</code>. The number of CPUs to use to estimate the ethnicity.
Default is <code>20</code>.</p></td>
    </tr>
    <tr>
      <th>dpi</th>
      <td><p>A <code>numeric</code>. The value for dpi when plotting the data.
Default is <code>120</code>.</p></td>
    </tr>
    <tr>
      <th>gg_fontsize</th>
      <td><p>A <code>numeric</code>. Value for the font size. Default is <code>12</code>.</p></td>
    </tr>
    <tr>
      <th>encoding</th>
      <td><p>A <code>character</code>. The encoding to be used for the html report.
Default is <code>"UTF-8"</code>.</p></td>
    </tr>
    </table>
    

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
                </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Mickaël Canouil, Mathilde Boissel, Lijiao Ning.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

